---
title: "DADA2_sludge_2023"
output: html_document
date: "2023-10-23"
editor_options: 
  chunk_output_type: console
---

## Acknowledgements

Thank you to our collaborator Dr. Erik Coats and our BGMP mentor Dr. Maxine Wren for technical support on this project. 
Data collection and analyses performed by the IIDS Genomics and Bioinformatics Resources Core at the University of Idaho were supported in part by NIH COBRE grant P30GM103324.

### Software Used

R 4.3.1 was used for all computations. The following packages were used:

| Package | Use |
|---------|-----|
| dplyr | general data processing |
| tibble | general data processing |
| tidyverse | general data processing | 
| DT | general data processing |
| devtools | general data processing | 
| ggplot2 | general data visualization |
| ggh4x | general data visualization |
| gridExtra | general data visualization | 
| ggpubr | general data visualization |
| viridis | colors for data visualization |
| rstatix | general data processing and visualization | 
| vegan | microbial community analysis |
| phyloseq | microbial community analysis | 
| microbiome | microbial community analysis |
| DESeq2 | normalization |
| MicEco | Euler graphs | 
| multcomp | multiple comparisons, analysis of variance | 
| ANCOMBC | differential abundance analysis | 
| DADA2 | resolving ASVs and assigning taxonomy | 
| fantaxtic | microbial abundance graphs |
| magrittr | depend for fantaxtic |
| ggnested | depend for fantaxtic |
| Biostrings | string container for ASVs | 
| DECIPHER | phylogenetic trees | 
| phangorn | phylogenetic trees |


## Data description

Activated sludge (AS) is a mixture of microbes that play an important part in waste water treatment (WWT). Living bacteria within AS remove polluting nutrients such as phosphorus and nitrogen before effluent is released back to the water cycle. This is an exploratory study focusing on taxonomic composition of these bacterial communities and their associated functional potential, suggesting the roles they may play in AS. In this project, we worked with data provided by Dr. Erik Coats and the Bioinformatics Core from the University of Idaho. The Coats lab has sampled AS from different waste water treatment setups: from a full-scale operational waste water treatment plant in Moscow, ID, to experimental sequencing batch reactors used for increasing PHA (polyhydroxyalkanoates) production which is a bioplastics precursor, to investigating more efficient EBPR (Enhanced Biological Phosphorus Removal) and nitrogen removal systems. 
  
The data span a range of years, and samples were collected from this variety of setups. Despite this variety, we can generally separate these systems by their intended specialized function: to remove phosphorus, to remove phosphorus and nitrogen, or to enhance PHA production.

| Variable | Description | 
| -------- | ----------- |
| System type | Scale and setup of system. Sequencing batch reactor, scale model WRRF, full scale WRRF. |
| Media type | Inflow media. Fermented dairy manure, synethetic wastewater, real wastewater. |
| Bin | Function of the system. Primarily phosphorus removal (P), primarily phosphorus + nitrogen removal (PN), primarily PHA production (PHA). |

88 individual samples of 16S rRNA gene were sequenced on an individual Illumina run using multiplexing. Raw reads were demultiplexed by the UI Core and this workflow begins with demultiplexed FASTQ files (read 1: R1 and read 2: R2). 

## Setup 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("dada2")
library("devtools")
library("phyloseq")
library("Biostrings")
library("ggplot2")
library("gridExtra")
library("dplyr")
library(tidyverse)
library(DECIPHER)
library(phangorn)
library(tibble)
library(ggh4x)
library(microbiome)
library(viridis)
library(ANCOMBC)
library(DT)
library(vegan)
library(multcomp)
library(fantaxtic)

```

```{r Set up directory}
path_R1 <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/R1_raw_reads" # CHANGE ME to the directory containing the fastq files after unzipping.
path_R2 <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/R2_raw_reads" # CHANGE ME to the directory containing the fastq files after unzipping.

working_directory <- "./files"

mylevels = c("Enrichment BSC", "Full Enrichment BSC", "Production Run BSC", "AE4","AE8", "AE12", "AE20", "BP_PX2", "BP_PX2.1", "BP_PX3", "BP_PX3.1", "BP_PX5", "BP_PX6", "FEBPR", "GEBPR", "NF", "SEBPR", "Winkler I", "Winkler J","MWWTP", "scalemodel")

```

```{r Extract sample names}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1.fastq and SAMPLENAME_R2.fastq
fnFs <- sort(list.files(path_R1, pattern="_R1.fastq", full.names = TRUE))
fnRs <- sort(list.files(path_R2, pattern="_R2.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

## Run DADA2 to resolve ASVs

```{r Inspect read quality profiles}
# Create the plots
plot1 <- plotQualityProfile(fnFs[1:2])
plot2 <- plotQualityProfile(fnRs[1:2])

# Combine plots into a single figure
grid.arrange(plot1, plot2, nrow = 1)

```

```{r Filter and trim}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path_R1, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path_R2, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(260,240), 
              maxN=0, maxEE=c(3,6), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

```{r Learn the Error Rates}

errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)

```

```{r Sample Inference}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
dadaFs[[1]]
```

```{r Merge paired reads}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=FALSE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

```{r Construct amplicon sequence variant (ASV) table}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

```

```{r Remove chimeras}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)
```

```{r Track reads through the pipeline}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

```{r Assign taxonomy}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/silva_nr99_v138.1_train_set.fa", multithread=TRUE)

taxa2 <- addSpecies(taxa, "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/silva_species_assignment_v138_1.fa")
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
``` 

```{r Transpose asv df to have ASVs as the rownames}
asv <- t(seqtab.nochim)
```

## Phyloseq

Phyloseq allows us to create a single object that holds taxonomic data, metadata, and ASV counts together.

```{r Setting up a dataframe for phyloseq}

library(readxl)
metadata <- as.data.frame(read_excel("./files/Illumina_Reads_updated_3_20_24.xlsx"))
colnames(metadata) <- gsub(" ", "_", colnames(metadata)) #replacing spaces in column names with underscores

# converting the Sample_ID column to the row names
meta <- metadata %>% column_to_rownames("Sample_ID")

# adding the sample_ID back into the df
meta$Sample_ID <- rownames(meta)

```

```{r Construct phyloseq object}
# filter out asvs with counts less than 1; these might be sequencing errors # 

asv = as.data.frame(asv) %>% mutate(count = rowSums(.)) %>% dplyr::select(count, everything()) %>% filter(count > 1) %>% dplyr::select(-count)

ps <- phyloseq(otu_table(asv, taxa_are_rows = TRUE), 
               sample_data(meta), 
               tax_table(taxa))

# save a phyloseq object to reset at later points# 
backup_phyobj_all = ps
```

```{r Renaming ASVs to short names}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))

```

## Data preprocessing

We can evaluate the overall alpha diversity of all the samples to get a general understanding of the spread of alpha diversity. Here are diversity values using richness, Shannon's, and Simpson's index calculations from the vegan package. 

```{r Overall alpha diversity evaluation by sample}
# prepare data into long format 
asv_df = as.data.frame(asv)
asv_names = tibble::rownames_to_column(asv_df, "ASV")
long_asv = pivot_longer(asv_names,
                        cols = 2:89,
                        names_to = "Group")
long_asv = dplyr::rename(long_asv, "name" = "ASV")
shared = long_asv

#alpha diversity functions
summary_shared = shared %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            n = sum(value))

p1 = summary_shared %>% 
  pivot_longer(cols = c(richness, shannon, simpson),
               names_to = "metric") %>% 
  ggplot(aes(x=Group, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 3, scales = "free_y")

p1

ggsave("summary_sample_diversity.png", plot=p1, path = "./files/images", width = 7, height = 9, units = "in")
```

To evaluate how sampling effort affects alpha diversity, we can use random subsampling. We have a few samples (under about 10,000) that show low sampling effort may affect their results. For the most part, points follow the flat line and show that we have sufficient sampling. We should consider a cutoff of 10,000.

```{r Rarefaction: evaluate how sample size affects diversity metrics}
#code from following the tutorial available at https://riffomonas.org/code_club/2022-03-17-alpha-diversity and  https://www.youtube.com/watch?v=wq1SXGQYgCs&ab_channel=RiffomonasProject
#generate random dataframe: random subsampling 

rand <- shared %>% 
  uncount(value) %>% 
  mutate(name = sample(name)) %>% 
  dplyr::count(Group, name, name = "value")

summary_rand = rand %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            n = sum(value))

p1= summary_rand %>% 
  pivot_longer(cols = c(richness, shannon, simpson),
               names_to = "metric") %>% 
  ggplot(aes(x=n, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 4, scales = "free_y")+
  ggtitle("Random sampling")+
  xlab("Sample size")

p1
#Save image
ggsave("summary_rarified_diversity.png", plot=p1, path = "./files/images", width = 7, height = 9, units = "in")
```

```{r data preprocessing}

phyobj_all = ps
phyobj_all

#Subset out other Kingdoms (one Eukaryote), mitochondria, and chloroplasts which are picked up by the Eubacteria primer

#prune to 10,000. See alpha diversity section.
samples = sample_sums(phyobj_all) > 10000
phyobj_all = prune_samples(samples, phyobj_all) 

phyobj_all = subset_taxa(phyobj_all, Kingdom == "Bacteria")
phyobj_all = subset_taxa(phyobj_all, Family != "Mitochondria")
phyobj_all = subset_taxa(phyobj_all, Order != "Chloroplast")

# List the samples that have been removed in preprocessing
new_table = otu_table(phyobj_all) %>% as.data.frame()
new_samples = colnames(new_table)
old_samples = rownames(meta)

diff = unique(old_samples[! old_samples %in% new_samples]) 
diff

phyobj_all

```

## Data preprocessing summary

We performed the following data preprocessing steps:
1a. Remove any reads not in the kingdom Bacteria
1b. Remove reads tagged as mitochondria
1c. Remove reads tagged as chloroplasts
2. Remove samples with less than 10,000 reads
3. Remove ASVs that only appear once. These are possible sequencing errors. 

```{r taxonomic resolution}
# Check for best resolution available from taxonomic assignment. #

#Before filtering
tax_df = tax_table(backup_phyobj_all) %>% as.data.frame()

sum(is.na(tax_df$Class))/length(tax_df$Class)
sum(is.na(tax_df$Order))/length(tax_df$Order)
sum(is.na(tax_df$Family))/length(tax_df$Family)
sum(is.na(tax_df$Genus))/length(tax_df$Genus)

# After filtering
tax_df = tax_table(phyobj_all) %>% as.data.frame()

sum(is.na(tax_df$Class))/length(tax_df$Class)
sum(is.na(tax_df$Order))/length(tax_df$Order)
sum(is.na(tax_df$Family))/length(tax_df$Family)
sum(is.na(tax_df$Genus))/length(tax_df$Genus)

# it looks like Family is the highest resolution possible, but we decided to go with Order for later analyses. # 
```

## Alpha diversity measures

```{r Plot diversity measures by media type, bin type, and system type with Shannon and Simpson}

p1 = plot_richness(ps,x = "Media_Type_1", measures=c("Shannon", "Simpson"), color="Media_Type_1")
p1
ggsave(filename = "shannon_simpson_mediatype.png", plot = p1, path = "./files/images", width = 7, height = 7, units = "in")

p1 = plot_richness(ps,x = "System_Type", measures=c("Shannon", "Simpson"), color="System_Type")
p1
ggsave(filename = "shannon_simpson_systemtype.png", plot = p1, path = "./files/images", width = 7, height = 7, units = "in")

p1 = plot_richness(ps,x = "Bin", measures=c("Shannon", "Simpson"), color="Bin")
p1
ggsave(filename = "shannon_simpson_bin.png", plot = p1, path = "./files/images", width = 7, height = 7, units = "in")
```

Here we use Shannon’s index, which considers both richness and evenness and tends to outperform Simpson’s index when many rarely occurring species are present, which is suitable for our dataset.

```{r Shannon's by Bin type}
#ANOVA
richness = phyobj_all %>% estimate_richness(measures = c("Shannon", "Simpson")) %>% rownames_to_column(var="Sample.ID")
  #subset meta 
samples = unique(richness$Sample.ID)
meta_subset = meta %>% rownames_to_column(var="Sample.ID") %>% filter(Sample.ID %in% samples)

diversity = full_join(meta_subset,richness, by="Sample.ID")

diversity$Bin = diversity$Bin %>% as.character() %>% as.factor()
shannon_bin = aov(Shannon ~ Bin - 1, data = diversity)

#Check Assumptions: ANOVA
qqnorm(shannon_bin$residuals)
qqline(shannon_bin$residuals)

data.frame(
  Media = diversity$Bin,
  Residuals = shannon_bin$residuals
) %>% 
  ggplot(aes(x = Media, y = Residuals, fill = Media)) +
  geom_boxplot()+
  ggtitle("Shannon's: Bin Type")

#Pairwise Comparisons with Tukey's Method
summary(glht(shannon_bin, linfct = mcp(Bin = "Tukey")))

```

```{r Alpha Diversity Analysis: figures}
library(ggpubr)
library(rstatix)

#Use rstatix package for plotting ggplot: slightly different results for Tukey Post-Hoc but same significance
#https://github.com/kassambara/ggpubr/issues/102#issuecomment-411897608

stat.test <- shannon_bin %>% tukey_hsd()
stat.test$custom.label <- ifelse(stat.test$p.adj <= 0.05, stat.test$p.adj, "ns")

p1 = plot_richness(phyobj_all, measures = c("Shannon"), x= "Bin")+
  xlab("") + theme(axis.text.x = element_text(angle = 0, hjust = 0.5, size = 12))+
    stat_pvalue_manual(stat.test, label = "p.adj.signif", y.position = c(6, 7, 8), tip.length = 0.01)

p1
#Save plot
ggsave(filename = "shannons_bin_anova.png", plot = p1, path = "./files/images", width = 7, height = 7, units = "in")
```

## Community structure with Euler diagrams 

```{r MicEco Euler diagrams, Full Sample}
library(MicEco)
library(gridExtra)

set.seed(130) #it looks really wonky otherwise.
euler1 = ps_euler(phyobj_all, "Media_Type_1", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#F2D000", "#F58609", "#A01892"))
p1 = grid.arrange(grobs = list(euler1), top = "Media Type, Full Sample")

set.seed(100)
euler2 = ps_euler(phyobj_all, "System_Type", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#347F92", "#A6CEE3","#94DE11" ))
p2 = grid.arrange(grobs = list(euler2), top = "System Type, Full Sample")

set.seed(125)
euler3 = ps_euler(phyobj_all, "Bin", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#347F92", "#A6CEE3","#94DE11" ))
p3 = grid.arrange(grobs = list(euler3), top = "Bin, Full Sample")

ggsave(filename = "euler_media.png", plot = p1, path = "./files/images", width = 7, height = 4, units = "in")
ggsave(filename = "euler_system.png", plot = p2, path = "./files/images", width = 7, height = 4, units = "in")
ggsave(filename = "euler_bin.png", plot = p3, path = "./files/images", width = 7, height = 4, units = "in")
```

## Beta Diversity analysis
  
We performed variance stabilization to correct for different sequencing depth or amplification bias before performing beta diversity analysis using the DESeq2 package. 

```{r preprocessing, variance stabilization (normalization)}

# Initial code courtesy of Dr. Maxine Wren, modified to fit our analysis

library(DESeq2)

#First, we need to create a DESeq2 object. Selected experimental variable is Bin (expected function)
diagdds <- phyloseq_to_deseq2(phyobj_all, ~Bin)

#Next, we need to estimate the variance between samples
diagdds <- estimateSizeFactors(diagdds,type="poscounts")
diagdds <- estimateDispersions(diagdds) 

#Now we need to use the variance between samples to normalize the data. 
#This reduces "noise" from variation between samples that may be due to differences
#in amplification or sequencing depth.
diagvst <- getVarianceStabilizedData(diagdds) 

#Now we need to make it BACK into an ASV/OTU table and slot it back into the phyloseq.
otu_table(phyobj_all) <- otu_table(diagvst, taxa_are_rows = TRUE)
ot_df <- data.frame(otu_table(phyobj_all))
ot_df[ot_df < 0.0] <- 0.0 # why this?
ot <- otu_table(ot_df, taxa_are_rows = TRUE)
sample_names(ot) <- sample_names(phyobj_all@otu_table)
otu_table(phyobj_all) <- otu_table(ot, taxa_are_rows = TRUE)

backup_var_phyobj_all = phyobj_all
```

```{r Functionality: Faprotax}
#Prepare a properly formatted tax table for use with Faprotax.

fasv = otu_table(phyobj_all) %>% as.data.frame() %>% rownames_to_column("otu")
ftax = tax_table(phyobj_all) %>% as.data.frame()

ftax$Kingdom = paste("d", ftax$Kingdom, sep = "__")
ftax$Phylum = paste("p", ftax$Phylum, sep = "__")
ftax$Class = paste("c", ftax$Class, sep = "__")
ftax$Order = paste("o", ftax$Order, sep = "__")
ftax$Family = paste("f", ftax$Family, sep = "__")
ftax$Genus = paste("g", ftax$Genus, sep = "__")
ftax$Species = paste("s", ftax$Species, sep = "__")

ftax$Taxonomy = paste(ftax$Kingdom, ftax$Phylum, ftax$Class, ftax$Order, ftax$Family, ftax$Genus, ftax$Species, sep = ";")

tax_table = ftax %>% rownames_to_column(var = "otu") %>% dplyr::select(otu, Taxonomy)


export_tax_table = dplyr::left_join(x = as.data.frame(tax_table),
                             y = as.data.frame(fasv),
                             by = "otu") %>% dplyr::select(-otu)
 
write_tsv(export_tax_table, "./files/silva_allgroups_taxtable_1-27-24-filt-norm.tsv")

#Run Faprotax on the command line. You will need numpy installed. 
# ./collapse_table.py -i silva_allgroups_taxtable_1-27-24-filt-norm.tsv -g FAPROTAX.txt -o silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv -r silva_allgroups_taxtable_1-27-24-filt-norm_report.txt -d "Taxonomy" -v

fapro_raw = read_tsv("./files/silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv", show_col_types = FALSE) %>% column_to_rownames(var = "group") %>% as.matrix()
```

## Relative Abundance at the Class level

We can visualize the diversity of each individual sample, each system, and each operation type at the class level. 

```{r Relative Abundance: by Class}

# Following the phyloseq relative abundance tutorial posted here: 
# https://github.com/joey711/phyloseq/issues/1701 

#subset only Bacteria
normalized_ps = phyobj_all

#select your variable of interest here 
merged_ps_phylum_SysName = merge_samples(normalized_ps, "System_Name") 
merged_ps_phylum_Bin = merge_samples(normalized_ps, "Bin")
merged_ps_phylum_Media = merge_samples(normalized_ps, "Media_Type_1")
merged_ps_phylum_SysType = merge_samples(normalized_ps, "System_Type")

#Merge at your taxonomic rank of interest, I chose Class here.
normalized_ps_phylum1 = tax_glom(normalized_ps, "Class", NArm = TRUE)
normalized_ps_phylum2 = tax_glom(merged_ps_phylum_SysName, "Class", NArm = TRUE)
normalized_ps_phylum3 = tax_glom(merged_ps_phylum_Bin, "Class", NArm = TRUE)
normalized_ps_phylum4 = tax_glom(merged_ps_phylum_Media, "Class", NArm = TRUE)
normalized_ps_phylum5 = tax_glom(merged_ps_phylum_SysType, "Class", NArm = TRUE)

# Transform counts to relative abundances
normalized_ps_phylum1_relabun = transform_sample_counts(normalized_ps_phylum1, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum2_relabun = transform_sample_counts(normalized_ps_phylum2, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum3_relabun = transform_sample_counts(normalized_ps_phylum3, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum4_relabun = transform_sample_counts(normalized_ps_phylum4, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum5_relabun = transform_sample_counts(normalized_ps_phylum5, function(OTU) OTU/sum(OTU) * 100)

#Get the number of unique colors and select that many from iWantHue https://medialab.github.io/iwanthue/ 
tax = taxa %>% as.data.frame()
length(unique(tax_df$Class))

myColors <- c("#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549")

# assign specific colors to variables
# https://stackoverflow.com/questions/6919025/how-to-assign-colors-to-categorical-variables-in-ggplot2-that-have-stable-mappin

ps@tax_table
tax_df = tax %>% as.data.frame()
names(myColors) <- levels(tax_df$Class)


p1 = psmelt(normalized_ps_phylum1_relabun) 
p1 = p1 %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Sample", x = "") +
        guides(fill = "none") + scale_fill_manual(name = "Class", values=myColors)

p2 = psmelt(normalized_ps_phylum2_relabun) 
p2$Sample = factor(p2$Sample, levels= mylevels)
p2 = p2 %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by System Name", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors)

p3 = psmelt(normalized_ps_phylum3_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Bin", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) + theme_bw() 

p4 = psmelt(normalized_ps_phylum4_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Media Type", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) + theme_bw() 

p5 = psmelt(normalized_ps_phylum5_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by System Type", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) + theme_bw()

p1
p2 
p3 
p4 
p5

#Save plots
ggsave(filename = "relabund_class_sampleid.png", plot = p1, path = "./files/images/relative_abundance", width = 7, height = 7, units = "in")
ggsave(filename = "relabund_class_samplename.png", plot = p2, path = "./files/images/relative_abundance", width = 7, height = 7, units = "in")
ggsave(filename = "relabund_class_bin.png", plot = p3, path = "./files/images/relative_abundance", width = 7, height = 7, units = "in")
ggsave(filename = "relabund_class_mediatype.png", plot = p4, path = "./files/images/relative_abundance", width = 7, height = 7, units = "in")
ggsave(filename = "relabund_class_systemtype.png", plot = p5, path = "./files/images/relative_abundance", width = 7, height = 7, units = "in")
```

We can see exactly what taxa are the most represented in these systems. 

```{r fantaxtic top genera}
# Top taxa package for plotting nested ggplots from a phyloseq object, fantaxtic, is here: 
# https://github.com/gmteunisse/Fantaxtic

#clear any merging and changing done to phyobj_all in last chunks
# phyobj_all = backup_var_phyobj_all

library(fantaxtic)
library(magrittr)
library(ggnested)
library(gridExtra)

#Find top 10 Orders across all samples
phyobj_all = backup_var_phyobj_all
top_asv <- top_taxa(phyobj_all, n_taxa = 10, tax_level = "Order")
p = plot_nested_bar(ps_obj = top_asv$ps_obj,
                top_level = "Class",
                nested_level = "Order") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black"))+
  xlab("") +
  ggtitle("Top 10 Orders across all samples")
p
ggsave(filename = "topfunction_order_all.png", plot = p, path = working_directory)

# Now, top 10 by separated by Bin
phyobj_all = backup_var_phyobj_all
norm_ps = merge_samples(phyobj_all, "Bin")
top_asv <- top_taxa(norm_ps, n_taxa = 10, tax_level = "Order")
p = plot_nested_bar(ps_obj = top_asv$ps_obj,
                top_level = "Class",
                nested_level = "Order")+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black"))+
  xlab("")+
  ggtitle("Top 10 Orders by Bin")
p
ggsave(filename = "topfunction_order_bin.png", plot = p, path = working_directory)

# Find top Classes
top_nested <- nested_top_taxa(phyobj_all,
                              top_tax_level = "Class",
                              nested_tax_level = "Order",
                              n_top_taxa = 3, 
                              n_nested_taxa = 3)
p = plot_nested_bar(ps_obj = top_nested$ps_obj,
                top_level = "Class",
                nested_level = "Order")+
  ggtitle("Top 3 orders of top 3 classes")
p
ggsave("top_3_class_order.png", plot = p, path = working_directory)

# Find top n by Variable Type 
top_grouped <- top_taxa(phyobj_all,
                        tax_level = "Order",
                        n_taxa = 5,
                        grouping = "Bin")
top_grouped$top_taxa %>% ggplot(aes(x=Bin, y = abundance*100, fill = Order)) +
  geom_bar(stat = "identity")+
  ylab("Percent abundance")

# change levels 

vars = get_variable(phyobj_all, "System_Name")
vars = factor(vars)
levels = levels(vars)
corlevels = c(levels[1:11], "Winkler I", "Winkler J", "Production Run BSC", "Full Enrichment BSC", "SEBPR", "FEBPR", "GEBPR","NF", "MWWTP", "scalemodel")
sample_data(phyobj_all)$System.Name = levels(corlevels)


# Facet wrap 
top_nested <- nested_top_taxa(phyobj_all,
                          top_tax_level = "Class",
                          nested_tax_level = "Order",
                          n_top_taxa = 10, 
                          n_nested_taxa = 1)
p = plot_nested_bar(top_nested$ps_obj,
                top_level = "Class",
                nested_level = "Order",
                legend_title = "Class and Order") +
  facet_wrap(~Bin,
             scales = "free_x",
             ncol = 6) +
  labs(title = "Relative abundances of the top 3 orders for each of the top 3 classes") +
  theme(plot.title = element_text(hjust = 0.5, 
                                  size = 20, 
                                  face = "bold"),
        legend.key.size = unit(10, 
                               "points"))
p
ggsave("top_3_classes_top5_order_bin.png", plot = p, path = working_directory, width = 8000, height = 4000, units = "px", dpi = 380)
```

```{r Top taxa poster figure generation}

# displays poorly in R markdown but saved image has correct dimensions #

#https://github.com/gmteunisse/Fantaxtic

library(extrafont)
library(showtext)
font_add(family = "skeena", regular = "\\\\wsl.localhost/Ubuntu/home/ruben/bioinfo/sludge/skeena_d.ttf")
showtext_auto()

#Find top 10 across all samples
norm_ps = merge_samples(phyobj_all, "Bin")
norm_ps@sam_data$Bin = factor(norm_ps@sam_data$Bin, levels = c("PN", "P", "PHA"))
top_asv <- top_taxa(norm_ps, n_taxa = 10, tax_level = "Order")

# options(repr.plot.width = 4, repr.plot.height =1.5) 
p3 = plot_nested_bar(ps_obj = top_asv$ps_obj,
                     top_level = "Class",
                     nested_level = "Order",
                     palette = c(Alphaproteobacteria = "#F3722C", 
                                 Bacteroidia = "#871762",
                                 Gammaproteobacteria = "#FFBB33"),
                     merged_clr = "#919191",
                     sample_order = c("PN", "P", "PHA"),
                     legend_title = "Orders by Class") +
  labs(title = "10 most represented orders") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black", family = "skeena", size = 40),
        axis.text.y = element_text(family = "skeena", size = 40),
        title = element_text(family = "skeena", size = 50),
        plot.background = element_rect(fill = "#F0EEE4"),
        plot.title = element_text(hjust = 0.5), 
        legend.background = element_blank(),
        legend.key = element_blank(),
        legend.title = element_text(family = "skeena", size = 40),
        legend.text = element_text(size = 40, family = "skeena")) +
  xlab("")
  


p3
ggsave("top_10_orders_bin.png", plot = p3, path = working_directory, width = 7, height = 4.5, units = "in", dpi = 300)
```

```{r ANCOMBC analysis}
# Code courtesy of Dr. Maxine Wren, modified for this data

# remove collinear variables from the metadata
filtered_meta = meta %>% dplyr::select(-c("System_Type", "Media_Type_1"))
ASV = otu_table(backup_phyobj_all)
TAX = tax_table(backup_phyobj_all)
SAMPLE = sample_data(filtered_meta)
filtered_ps = phyloseq(ASV, TAX, SAMPLE)

## remove any taxa with <10 reads in the total data set (sum of all samples <10). This is because extremely rare taxa can produce a series of false positives. 

filtered_ps = prune_taxa(taxa_sums(filtered_ps) > 10, filtered_ps) 

#check the results of filtering
otu_df = otu_table(filtered_ps) %>% as.data.frame() %>% mutate(sums = rowSums(.)) %>% dplyr::select(sums, everything())

# Continuous covariate: "Collection.Date"
# Categorical covariates: “System.Name, Bin”
# The group variable of interest: “Bin”
# Three groups: "P removal", "N and P removal", 
# The reference group: “Full scale WRRF”

# refactor the categorical variables so PHA bioreactor is the comparator
sample_data(filtered_ps)$Bin <- factor(sample_data(filtered_ps)$Bin,
                                      levels=c("PHA bioplastics reactor", 
                                               "N and P removal", 
                                               "P removal"))
# If I do the fix_formula as System.Name + Collection.Date + Bin it gives me an error about checking for collinearity or missing values in my covariates, so I am just running it with "bin" as the formula.

ancombc_out = ancombc2(filtered_ps,
                             fix_formula = "Bin",
                             tax_level = "Order",
                             group = "Bin",
                             struc_zero = TRUE,
                             neg_lb = TRUE,
                             verbose = TRUE,
                             alpha = 0.05,
                             global = TRUE,
                             pairwise = TRUE,
                             dunnet = TRUE
                             )
```

```{r taxon presence/absence}
# taxon presence/absence
tab_zero = ancombc_out$zero_ind %>% column_to_rownames(var = "taxon") #%>% mutate(sum = rowSums(.)) %>% filter(sum == 2)

res_prim = ancombc_out$res
res_global= ancombc_out$res_global %>% mutate_if(is.numeric, function(x) round(x, 5))
res_global_passed = res_global %>% filter(passed_ss == TRUE) %>% filter(diff_abn == TRUE)
diff_taxons = res_global_passed$taxon
diff_taxons
```

We can also visualize the diversity of functions present in each system. 

```{r Relative Abundance: by functionality}

# Replace with Farris stuff here # 

#Filter out subgroupings of functionality
# fapro_filt = fapro_raw %>% t() %>% as.data.frame() %>%  dplyr::select(-c(oxygenic_photoautotrophy, methanol_oxidation, dark_sulfide_oxidation, aerobic_nitrite_oxidation, human_gut, human_pathogens_all, mammal_gut, anoxygenic_photoautotrophy, anoxygenic_photoautotrophy_S_oxidizing, photosynthetic_cyanobacteria, photoheterotrophy, thiosulfate_respiration, methanotrophy, aliphatic_non_methane_hydrocarbon_degradation, iron_respiration, aerobic_chemoheterotrophy, nitrite_respiration, photoautotrophy, sulfur_respiration, nitrate_denitrification, nitrite_denitrification, nitrous_oxide_denitrification, denitrification, nitrate_respiration, nitrogen_respiration, human_associated, aromatic_hydrocarbon_degradation, chloroplasts)) %>% t()

# func_ASV = otu_table(fapro_raw, taxa_are_rows = TRUE)
# SAMPLE = sample_data(meta)
# phyobj_func = phyloseq(func_ASV, SAMPLE)
# 
# # plot all functions by each individual sample
# merged_ps_phylum_SysName = merge_samples(phyobj_func, "System.Name") 
# ps_phylum1_relabun = transform_sample_counts(merged_ps_phylum_SysName, function(OTU) OTU/sum(OTU) * 100)
# myColors <- c("#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549","#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549")
# p = psmelt(ps_phylum1_relabun) %>%
#         ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
#         geom_bar(stat = "identity") +
#         theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#         labs(y = "Relative Abundance", title = "Relative abundances of tagged functions", x = "") + 
#         guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) 
# p
# ggsave(filename = "allfunctions_systemname.png", plot = p, path = working_directory)
# 
# # get top functions for bins
# 
# func_ASV = otu_table(fapro_raw, taxa_are_rows = TRUE)
# phyobj_func = phyloseq(func_ASV, SAMPLE)
# 
# ps = phyobj_func
# 
# p = subset_samples(ps, Bin == "P")
# p10 = names(sort(taxa_sums(p), TRUE)[1:10])
# pn = subset_samples(ps, Bin == "PN")
# pn10 = names(sort(taxa_sums(pn), TRUE)[1:10])
# pha = subset_samples(ps, Bin == "PHA")
# pha10 = r10 = names(sort(taxa_sums(pha), TRUE)[1:10])
# top10 = unique(c(pha10, p10, pn10))
# 
# ps = prune_species(top10, ps)
# 
# merged_ps = merge_samples(ps, "Bin") #change what variable here
# 
# merged_ps = transform_sample_counts(merged_ps, function(OTU) OTU/sum(OTU) * 100)
# 
# #get the number of unique functions and select that many colors as above
# 
# myColors <- c("#006403", "#9d55dd", "#d5b500", "#88258e", "#a9d382", "#b0008a", "#d04b00", "#0082d2", "#ff635a", "#e99eff", "#814030", "#fcb0d3", "#9c1b5c", "firebrick", "#a0029b", "seagreen")
# # tax_df = otu_table(ps) %>% as.data.frame()
# # names(myColors) <- rownames(tax_df)
# 
# p1 = psmelt(merged_ps) %>% 
#   ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
#         geom_bar(stat = "identity") +
#         theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#         labs(y = "Relative Abundance", title = "Most common functions", x = "") +
#         scale_fill_manual(values=myColors) + 
#         guides(fill=guide_legend(title="Function")) + theme_bw()
# p1
# ggsave(filename = "topfunction_bin.png", plot = p1, path = working_directory)
```


# Bray Curtis NMDS 

```{r variance stabilized NMDS: Bin whole sample}
set.seed(100)
#Bray-Curtis distance matrix calculation by System Type

physeq_ord <- ordinate(
 physeq = phyobj_all,
 method = "NMDS",
 formula = ~Bin,
 distance = "bray"
)
stress = as.character(round(physeq_ord$stress, digits = 3))
stresslab = paste("Stress =", stress)

p1 = plot_ordination(
 physeq = phyobj_all,
 ordination = physeq_ord,
 color = "Bin"
) +
 aes(group = Bin) +
 geom_point(aes(color = Bin), alpha = 0.95, size = 6) +
  scale_color_manual(values=c("#A6CEE3","#347F92","#94DE11"))+
 stat_ellipse(type = "t", geom = "polygon", position = "identity", alpha = .05, aes(fill = Bin)) + 
 scale_fill_manual(values=c("#A6CEE3","#347F92","#94DE11")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
    panel.background = element_blank(), axis.line = element_line(colour = "black"),strip.text = element_text(colour = 'white'),text = element_text(size=12),legend.text=element_text(size=12),legend.title = element_blank())+
  annotate("text", x = 2, y = -2, label = stresslab) + 
  annotate("text", x = 2, y = -2.5, label = "Non-metric fit, R² = 0.985")

p1
ggsave("nmds_Bin.png", plot = p1, path=working_directory)

#filter out ASVs that have only 0 values
ot_df = otu_table(phyobj_all) %>% as.data.frame()

#247 -> 239
ot_df = ot_df %>% mutate(sum = rowSums(.)) %>% dplyr::select(sum, everything()) %>% filter(sum != 0) %>% dplyr::select(-sum) %>% t()
samples = rownames(ot_df)
adonis_dist = vegdist(ot_df, method = "bray") 

# adonis_mds = metaMDS(adonis_dist) %>% scores() %>% 
#   as_tibble(rownames="Group")

#subset samples needed from meta
# needed_samples = rownames(ot_df)
# parcu_adonis_meta = meta %>% rownames_to_column(var="Group") %>% filter(Group %in% needed_samples)
# 
# meta_nmds = inner_join(adonis_mds, parcu_adonis_meta, by = "Group")

# centroid_systemtype = meta_nmds %>% 
#   group_by(Bin) %>% 
#   summarize(NMDS1 = mean(NMDS1), NMDS2 = mean(NMDS2))
# 
# meta_nmds %>% 
#   ggplot(aes(x=NMDS1, y = NMDS2, color = Bin))+
#   geom_point() +
#   stat_ellipse(show.legend = FALSE) +
#   geom_point(data = centroid_systemtype, size = 2, shape = 21, color = "black", aes(fill=Bin), show.legend = FALSE)+
#  # ggtitle("Media Type, bray-curtis")+
#   scale_colour_discrete(name="Bin") +
#   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
#     panel.background = element_blank(), axis.line = element_line(colour = "black"))

stressplot(physeq_ord)

submeta = meta %>% t() %>% as.data.frame %>% dplyr::select(any_of(samples)) %>% t() %>% as.data.frame()

# PERMANOVA # 
bintype = adonis2(adonis_dist ~ Bin, data = submeta, permutations = 1e4)
bintype$`Pr(>F)`[1] #p-value 

bd = betadisper(adonis_dist, submeta$Bin)
anova(bd)


```

```{r NMDS by bin, for poster}
# displays poorly in R but saves image with correct dimensions # 

physeq_ord <- ordinate(
 physeq = phyobj_all,
 method = "NMDS",
 formula = ~Bin,
 distance = "bray"
)

stress = as.character(round(physeq_ord$stress, digits = 3))
stresslab = paste("Stress =", stress)

p1 = plot_ordination(
 physeq = phyobj_all,
 ordination = physeq_ord,
 color = "Bin"
) +
 aes(group = Bin) +
 geom_point(aes(color = Bin), alpha = 0.95, size = 5) +
  scale_color_manual(values=c("#F3722C","#871762","#90B30E"))+
  scale_fill_manual(values=c("#F3722C","#871762","#90B30E")) +
  stat_ellipse(type = "t", geom = "polygon", position = "identity", alpha = .05, aes(fill = Bin)) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
    panel.background = element_blank(), axis.line = element_line(colour = "black"),strip.text = element_text(colour = 'white'),text = element_text(size=60, family = "skeena"),legend.text=element_text(size=60),legend.title = element_blank())+
  annotate("text", x = 4.2, y = -2, label = stresslab, size = 12, family = "skeena")+
  annotate("text", x = 3.6, y = -2.5, label = "Non-metric fit, R² = 0.985", size = 12, family = "skeena")+
  theme(plot.background = element_rect(fill = "#F0EEE4"), legend.background = element_rect(fill = "#F0EEE4"))
  

p1
ggsave("nmds_Bin.png", plot = p1, path=working_directory, height = 5, width = 7, dpi = 300)
```

Efforts were made at the beginning of the project to investigate Parcubacteria specifically. Beta diversity analysis using NMDS proved difficult due to the sparsity of Parcubacteria differences between samples. Presence-absence and abundance testing was possible, however, and revealed that PN systems and systems with real wastewater had a greater amount and variety of Parcubacteria.


=======
---
title: "DADA2_sludge_2023"
output: html_document
date: "2023-10-23"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("dada2")
library("devtools")
library("phyloseq")
library("Biostrings")
library("ggplot2")
library("gridExtra")
library("dplyr")
library(tidyverse)
library(DECIPHER)
library(phangorn)
library(tibble)
library(ggh4x)
library(microbiome)
library(viridis)
library(ANCOMBC)
library(DT)
library(vegan)
library(multcomp)
library(fantaxtic)

```


```{r Set up directory}
path_R1 <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/R1_raw_reads" # CHANGE ME to the directory containing the fastq files after unzipping.
path_R2 <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/R2_raw_reads" # CHANGE ME to the directory containing the fastq files after unzipping.

working_directory <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/"

mylevels = c("Enrichment BSC", "Full Enrichment BSC", "Production Run BSC", "AE4","AE8", "AE12", "AE20", "BP_PX2", "BP_PX2.1", "BP_PX3", "BP_PX3.1", "BP_PX5", "BP_PX6", "FEBPR", "GEBPR", "NF", "SEBPR", "Winkler I", "Winkler J","MWWTP", "scalemodel")

```

```{r Extract sample names}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1.fastq and SAMPLENAME_R2.fastq
fnFs <- sort(list.files(path_R1, pattern="_R1.fastq", full.names = TRUE))
fnRs <- sort(list.files(path_R2, pattern="_R2.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

Ruben starts after the asv is made 

```{r Inspect read quality profiles}
# Create the plots
plot1 <- plotQualityProfile(fnFs[1:2])
plot2 <- plotQualityProfile(fnRs[1:2])

# Combine plots into a single figure
grid.arrange(plot1, plot2, nrow = 1)

```

```{r Filter and trim}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path_R1, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path_R2, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(260,240), 
              maxN=0, maxEE=c(3,6), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

```{r Learn the Error Rates}

errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)

```

```{r Sample Inference}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
dadaFs[[1]]
```

```{r Merge paired reads}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=FALSE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

```{r Construct amplicon sequence variant (ASV) table}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

```

```{r Remove chimeras}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)
```

```{r Track reads through the pipeline}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

```{r Assign taxonomy}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/silva_nr99_v138.1_train_set.fa", multithread=TRUE)

taxa2 <- addSpecies(taxa, "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/silva_species_assignment_v138_1.fa")
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
``` 

```{r Transpose asv df to have ASVs as the rownames}
asv <- t(seqtab.nochim)
```

# Phyloseq

```{r Setting up a dataframe for phyloseq}

library(readxl)
metadata <- as.data.frame(read_excel("/Users/leycufur/bioinfo/Bi624/Sludge/Illumina_Reads_updated_1-10-24.xlsx"))
colnames(metadata) <- gsub(" ", "_", colnames(metadata)) #replacing spaces in column names with underscores

# converting the Sample_ID column to the row names
meta <- metadata %>% column_to_rownames("Sample_ID")

# adding the sample_ID back into the df
meta$Sample_ID <- rownames(meta)

```

```{r Construct phyloseq object}
asv = as.data.frame(asv) %>% mutate(count = rowSums(.)) %>% dplyr::select(count, everything()) %>% filter(count > 1) %>% dplyr::select(-count)

ps <- phyloseq(otu_table(asv, taxa_are_rows = TRUE), 
               sample_data(meta), 
               tax_table(taxa))
backup_phyobj_all = ps
```

```{r Renaming ASVs to short names}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))

```

```{r Plot richness by media type with Shannon and Simpson}
plot_richness(ps,x = "Media_Type_1", measures=c("Shannon", "Simpson"), color="Media_Type_1")

```

```{r Overall alpha diversity evaluation by sample}
# prepare data into long format 
asv_df = as.data.frame(asv)
asv_names = tibble::rownames_to_column(asv_df, "ASV")
long_asv = pivot_longer(asv_names,
                        cols = 2:89,
                        names_to = "Group")
long_asv = dplyr::rename(long_asv, "name" = "ASV")
shared = long_asv

#alpha diversity functions
summary_shared = shared %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            n = sum(value))

p1 = summary_shared %>% 
  pivot_longer(cols = c(richness, shannon, simpson),
               names_to = "metric") %>% 
  ggplot(aes(x=Group, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 3, scales = "free_y")

p1

ggsave("summary_sample_diversity.png", plot=p1, path = working_directory)
```

```{r Rarefaction: evaluate how sample size affects diversity metrics}
#code from following the tutorial available at https://riffomonas.org/code_club/2022-03-17-alpha-diversity and  https://www.youtube.com/watch?v=wq1SXGQYgCs&ab_channel=RiffomonasProject
#generate random dataframe: random subsampling 

rand <- shared %>% 
  uncount(value) %>% 
  mutate(name = sample(name)) %>% 
  dplyr::count(Group, name, name = "value")

summary_rand = rand %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            n = sum(value))

p1= summary_rand %>% 
  pivot_longer(cols = c(richness, shannon, simpson),
               names_to = "metric") %>% 
  ggplot(aes(x=n, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 4, scales = "free_y")+
  ggtitle("Random sampling")+
  xlab("Sample size")

p1
#Save image
ggsave("summary_rarified_diversity.png", plot=p1, path = working_directory)
```

```{r data preprocessing}

phyobj_all = ps
phyobj_all

#Subset out other Kingdoms (one Eukaryote), mitochondria, and chloroplasts which are picked up by the eubacteria primer


#prune to 10,000. See alpha diversity section.
samples = sample_sums(phyobj_all) > 10000
phyobj_all = prune_samples(samples, phyobj_all) 

phyobj_all = subset_taxa(phyobj_all, Kingdom == "Bacteria")
phyobj_all = subset_taxa(phyobj_all, Family != "Mitochondria")
phyobj_all = subset_taxa(phyobj_all, Order != "Chloroplast")


# List the samples that have been removed in preprocessing
new_table = otu_table(phyobj_all) %>% as.data.frame()
new_samples = colnames(new_table)
old_samples = rownames(meta)

diff = unique(old_samples[! old_samples %in% new_samples]) 
diff

phyobj_all

```

## Data preprocessing 

We performed the following data preprocessing steps:
1a. Remove any reads not in the kingdom Bacteria
1b. Remove reads tagged as mitochondria
1c. Remove reads tagged as chloroplasts
2. Remove samples with less than 10,000 reads
3. Remove ASVs that only appear once. These are possible sequencing errors. 

```{r taxonomic resolution}
# Check for best resolution available from taxonomic assignment. #

#Before filtering
tax_df = tax_table(backup_phyobj_all) %>% as.data.frame()

sum(is.na(tax_df$Class))/length(tax_df$Class)
sum(is.na(tax_df$Order))/length(tax_df$Order)
sum(is.na(tax_df$Family))/length(tax_df$Family)
sum(is.na(tax_df$Genus))/length(tax_df$Genus)

# After filtering
tax_df = tax_table(phyobj_all) %>% as.data.frame()

sum(is.na(tax_df$Class))/length(tax_df$Class)
sum(is.na(tax_df$Order))/length(tax_df$Order)
sum(is.na(tax_df$Family))/length(tax_df$Family)
sum(is.na(tax_df$Genus))/length(tax_df$Genus)
```

```{r Shannon's by Bin type, fig.show = "hide"}
#ANOVA
richness = phyobj_all %>% estimate_richness(measures = c("Shannon", "Simpson")) %>% rownames_to_column(var="Sample.ID")
  #subset meta 
samples = unique(richness$Sample.ID)
meta_subset = meta %>% rownames_to_column(var="Sample.ID") %>% filter(Sample.ID %in% samples)

diversity = full_join(meta_subset,richness, by="Sample.ID")

diversity$Bin = diversity$Bin %>% as.character() %>% as.factor()
shannon_bin = aov(Shannon ~ Bin - 1, data = diversity)

#Check Assumptions: ANOVA
qqnorm(shannon_bin$residuals)
qqline(shannon_bin$residuals)

data.frame(
  Media = diversity$Bin,
  Residuals = shannon_bin$residuals
) %>% 
  ggplot(aes(x = Media, y = Residuals, fill = Media)) +
  geom_boxplot()+
  ggtitle("Shannon's: Bin Type")

#Pairwise Comparisons with Tukey's Method
summary(glht(shannon_bin, linfct = mcp(Bin = "Tukey")))

```

```{r Alpha Diversity Analysis: figures}
library(ggpubr)
library(rstatix)

#Use rstatix package for plotting ggplot: slightly different results for Tukey Post-Hoc but same significance
#https://github.com/kassambara/ggpubr/issues/102#issuecomment-411897608

stat.test <- shannon_bin %>% tukey_hsd()
stat.test$custom.label <- ifelse(stat.test$p.adj <= 0.05, stat.test$p.adj, "ns")

p1 = plot_richness(phyobj_all, measures = c("Shannon"), x= "Bin")+
  xlab("") + theme(axis.text.x = element_text(angle = 0, hjust = 0.5, size = 12))+
    stat_pvalue_manual(stat.test, label = "p.adj.signif", y.position = c(6, 7, 8), tip.length = 0.01)

p1
#Save plot
ggsave(filename = "shannons_bin.png", plot = p1, path = working_directory)
```

## Community structure with Euler diagrams 
```{r MicEco Euler diagrams, Full Sample}
library(MicEco)
library(gridExtra)

set.seed(130) #it looks really wonky otherwise.
euler1 = ps_euler(phyobj_all, "Media_Type_1", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#F2D000", "#F58609", "#A01892"))
p1 = grid.arrange(grobs = list(euler1), top = "Media Type, Full Sample")

set.seed(100)
euler2 = ps_euler(phyobj_all, "System_Type", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#347F92", "#A6CEE3","#94DE11" ))
p2 = grid.arrange(grobs = list(euler2), top = "System Type, Full Sample")

set.seed(123)
euler3 = ps_euler(phyobj_all, "Bin", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#347F92", "#A6CEE3","#94DE11" ))
p3 = grid.arrange(grobs = list(euler3), top = "Bin, Full Sample")

ggsave(filename = "euler_media.png", plot = p1, path = working_directory)
ggsave(filename = "euler_system.png", plot = p2, path = working_directory)
ggsave(filename = "euler_bin.png", plot = p3, path = working_directory)
```

## Beta Diversity analysis
  
We performed variance stabilization to correct for different sequencing depth or amplification bias before performing beta diversity analysis using the DESeq2 package. 

```{r preprocessing, variance stabilization (normalization)}

# Initial code courtesy of Dr. Maxine Wren, modified to fit our analysis

library(DESeq2)

#First, we need to create a DESeq2 object. Selected experimental variable is Bin (expected function)
diagdds <- phyloseq_to_deseq2(phyobj_all, ~Bin)

#Next, we need to estimate the variance between samples
diagdds <- estimateSizeFactors(diagdds,type="poscounts")
diagdds <- estimateDispersions(diagdds) 

#Now we need to use the variance between samples to normalize the data. 
#This reduces "noise" from variation between samples that may be due to differences
#in amplification or sequencing depth.
diagvst <- getVarianceStabilizedData(diagdds) 

#Now we need to make it BACK into an ASV/OTU table and slot it back into the phyloseq.
otu_table(phyobj_all) <- otu_table(diagvst, taxa_are_rows = TRUE)
ot_df <- data.frame(otu_table(phyobj_all))
ot_df[ot_df < 0.0] <- 0.0 # why this?
ot <- otu_table(ot_df, taxa_are_rows = TRUE)
sample_names(ot) <- sample_names(phyobj_all@otu_table)
otu_table(phyobj_all) <- otu_table(ot, taxa_are_rows = TRUE)

backup_var_phyobj_all = phyobj_all
```

```{r Functionality: Faprotax}
#Prepare a properly formatted tax table for use with Faprotax.

fasv = otu_table(phyobj_all) %>% as.data.frame() %>% rownames_to_column("otu")
ftax = tax_table(phyobj_all) %>% as.data.frame()

ftax$Kingdom = paste("d", ftax$Kingdom, sep = "__")
ftax$Phylum = paste("p", ftax$Phylum, sep = "__")
ftax$Class = paste("c", ftax$Class, sep = "__")
ftax$Order = paste("o", ftax$Order, sep = "__")
ftax$Family = paste("f", ftax$Family, sep = "__")
ftax$Genus = paste("g", ftax$Genus, sep = "__")
ftax$Species = paste("s", ftax$Species, sep = "__")

ftax$Taxonomy = paste(ftax$Kingdom, ftax$Phylum, ftax$Class, ftax$Order, ftax$Family, ftax$Genus, ftax$Species, sep = ";")

tax_table = ftax %>% rownames_to_column(var = "otu") %>% dplyr::select(otu, Taxonomy)


export_tax_table = dplyr::left_join(x = as.data.frame(tax_table),
                             y = as.data.frame(fasv),
                             by = "otu") %>% dplyr::select(-otu)
 
write_tsv(export_tax_table, "silva_allgroups_taxtable_1-27-24-filt-norm.tsv")

#Run Faprotax on the command line. You will need numpy installed. 
# ./collapse_table.py -i silva_allgroups_taxtable_1-27-24-filt-norm.tsv -g FAPROTAX.txt -o silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv -r silva_allgroups_taxtable_1-27-24-filt-norm_report.txt -d "Taxonomy" -v

fapro_raw = read_tsv("\\\\wsl.localhost/Ubuntu/home/ruben/bioinfo/sludge/silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv", show_col_types = FALSE) %>% column_to_rownames(var = "group") %>% as.matrix()
```

## Relative Abundance

We can visualize the diversity of each individual sample, each system, and each operation type at the class level. 

```{r Relative Abundance: by Class}

# Following the phyloseq relative abundance tutorial posted here: 
# https://github.com/joey711/phyloseq/issues/1701 

#subset only Bacteria
normalized_ps = phyobj_all

#select your variable of interest here 
merged_ps_phylum_SysName = merge_samples(normalized_ps, "System_Name") 
merged_ps_phylum_Bin = merge_samples(normalized_ps, "Bin")
# merged_ps_phylum_Media = merge_samples(normalized_ps, "Media.Type.1") 
# merged_ps_phylum_SysType = merge_samples(normalized_ps, "System.Type") 

#Merge at your taxonomic rank of interest, I chose Class here.
normalized_ps_phylum1 = tax_glom(normalized_ps, "Class", NArm = TRUE)
normalized_ps_phylum2 = tax_glom(merged_ps_phylum_SysName, "Class", NArm = TRUE)
normalized_ps_phylum3 = tax_glom(merged_ps_phylum_Bin, "Class", NArm = TRUE)
# normalized_ps_phylum4 = tax_glom(merged_ps_phylum_Media, "Class", NArm = TRUE)
# normalized_ps_phylum5 = tax_glom(merged_ps_phylum_SysType, "Class", NArm = TRUE)

# Transform counts to relative abundances
normalized_ps_phylum1_relabun = transform_sample_counts(normalized_ps_phylum1, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum2_relabun = transform_sample_counts(normalized_ps_phylum2, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum3_relabun = transform_sample_counts(normalized_ps_phylum3, function(OTU) OTU/sum(OTU) * 100)
# normalized_ps_phylum4_relabun = transform_sample_counts(normalized_ps_phylum4, function(OTU) OTU/sum(OTU) * 100)
# normalized_ps_phylum5_relabun = transform_sample_counts(normalized_ps_phylum5, function(OTU) OTU/sum(OTU) * 100)

#Get the number of unique colors and select that many from iWantHue https://medialab.github.io/iwanthue/ 
tax = taxa %>% as.data.frame()
length(unique(tax_df$Class))

myColors <- c("#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549")

# assign specific colors to variables
# https://stackoverflow.com/questions/6919025/how-to-assign-colors-to-categorical-variables-in-ggplot2-that-have-stable-mappin


ps@tax_table
tax_df = tax %>% as.data.frame()
names(myColors) <- levels(tax_df$Class)


p1 = psmelt(normalized_ps_phylum1_relabun) 
p1 = p1 %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Sample", x = "") +
        guides(fill = "none") + scale_fill_manual(name = "Class", values=myColors)

p2 = psmelt(normalized_ps_phylum2_relabun) 
p2$Sample = factor(p2$Sample, levels= mylevels)
p2 = p2 %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by System", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors)

p3 = psmelt(normalized_ps_phylum3_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Bin", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) 


p1 
p2 
p3 + theme_bw()

#Save plots
ggsave(filename = "relabund_class_sampleid.png", plot = p1, path = working_directory)
ggsave(filename = "relabund_class_samplename.png", plot = p2, path = working_directory)
```

We can also visualize the diversity of functions present in each system. 

```{r Relative Abundance: by functionality}
#Filter out subgroupings of functionality
# fapro_filt = fapro_raw %>% t() %>% as.data.frame() %>%  dplyr::select(-c(oxygenic_photoautotrophy, methanol_oxidation, dark_sulfide_oxidation, aerobic_nitrite_oxidation, human_gut, human_pathogens_all, mammal_gut, anoxygenic_photoautotrophy, anoxygenic_photoautotrophy_S_oxidizing, photosynthetic_cyanobacteria, photoheterotrophy, thiosulfate_respiration, methanotrophy, aliphatic_non_methane_hydrocarbon_degradation, iron_respiration, aerobic_chemoheterotrophy, nitrite_respiration, photoautotrophy, sulfur_respiration, nitrate_denitrification, nitrite_denitrification, nitrous_oxide_denitrification, denitrification, nitrate_respiration, nitrogen_respiration, human_associated, aromatic_hydrocarbon_degradation, chloroplasts)) %>% t()

func_ASV = otu_table(fapro_raw, taxa_are_rows = TRUE)
SAMPLE = sample_data(meta)
phyobj_func = phyloseq(func_ASV, SAMPLE)

# plot all functions by each individual sample
merged_ps_phylum_SysName = merge_samples(phyobj_func, "System.Name") 
ps_phylum1_relabun = transform_sample_counts(merged_ps_phylum_SysName, function(OTU) OTU/sum(OTU) * 100)
myColors <- c("#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549","#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549")
p = psmelt(ps_phylum1_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Relative abundances of tagged functions", x = "") + 
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) 
p
ggsave(filename = "allfunctions_systemname.png", plot = p, path = working_directory)

# get top functions for bins

func_ASV = otu_table(fapro_raw, taxa_are_rows = TRUE)
phyobj_func = phyloseq(func_ASV, SAMPLE)

ps = phyobj_func

p = subset_samples(ps, Bin == "P")
p10 = names(sort(taxa_sums(p), TRUE)[1:10])
pn = subset_samples(ps, Bin == "PN")
pn10 = names(sort(taxa_sums(pn), TRUE)[1:10])
pha = subset_samples(ps, Bin == "PHA")
pha10 = r10 = names(sort(taxa_sums(pha), TRUE)[1:10])
top10 = unique(c(pha10, p10, pn10))

ps = prune_species(top10, ps)

merged_ps = merge_samples(ps, "Bin") #change what variable here

merged_ps = transform_sample_counts(merged_ps, function(OTU) OTU/sum(OTU) * 100)

#get the number of unique functions and select that many colors as above

myColors <- c("#006403", "#9d55dd", "#d5b500", "#88258e", "#a9d382", "#b0008a", "#d04b00", "#0082d2", "#ff635a", "#e99eff", "#814030", "#fcb0d3", "#9c1b5c", "firebrick", "#a0029b", "seagreen")
# tax_df = otu_table(ps) %>% as.data.frame()
# names(myColors) <- rownames(tax_df)

p1 = psmelt(merged_ps) %>% 
  ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Most common functions", x = "") +
        scale_fill_manual(values=myColors) + 
        guides(fill=guide_legend(title="Function")) + theme_bw()
p1
ggsave(filename = "topfunction_bin.png", plot = p1, path = working_directory)
```
We can see exactly what taxa are the most represented in these systems. 

```{r fantaxtic top genera}
# Top taxa package for plotting nested ggplots from a phyloseq object, fantaxtic, is here: 
# https://github.com/gmteunisse/Fantaxtic

#clear any merging and changing done to phyobj_all in last chunks
# phyobj_all = backup_var_phyobj_all

library(fantaxtic)
library(magrittr)
library(ggnested)
library(gridExtra)

#Find top 10 Orders across all samples
phyobj_all = backup_var_phyobj_all
top_asv <- top_taxa(phyobj_all, n_taxa = 10, tax_level = "Order")
p = plot_nested_bar(ps_obj = top_asv$ps_obj,
                top_level = "Class",
                nested_level = "Order") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black"))+
  xlab("") +
  ggtitle("Top 10 Orders across all samples")
p
ggsave(filename = "topfunction_order_all.png", plot = p, path = working_directory)

# Now, top 10 by separated by Bin
phyobj_all = backup_var_phyobj_all
norm_ps = merge_samples(phyobj_all, "Bin")
top_asv <- top_taxa(norm_ps, n_taxa = 10, tax_level = "Order")
p = plot_nested_bar(ps_obj = top_asv$ps_obj,
                top_level = "Class",
                nested_level = "Order")+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black"))+
  xlab("")+
  ggtitle("Top 10 Orders by Bin")
p
ggsave(filename = "topfunction_order_bin.png", plot = p, path = working_directory)

# Find top Classes
top_nested <- nested_top_taxa(phyobj_all,
                              top_tax_level = "Class",
                              nested_tax_level = "Order",
                              n_top_taxa = 3, 
                              n_nested_taxa = 3)
p = plot_nested_bar(ps_obj = top_nested$ps_obj,
                top_level = "Class",
                nested_level = "Order")+
  ggtitle("Top 3 orders of top 3 classes")
p
ggsave("top_3_class_order.png", plot = p, path = working_directory)

# Find top n by Variable Type 
top_grouped <- top_taxa(phyobj_all,
                        tax_level = "Order",
                        n_taxa = 5,
                        grouping = "Bin")
top_grouped$top_taxa %>% ggplot(aes(x=Bin, y = abundance*100, fill = Order)) +
  geom_bar(stat = "identity")+
  ylab("Percent abundance")

# change levels 

vars = get_variable(phyobj_all, "System_Name")
vars = factor(vars)
levels = levels(vars)
corlevels = c(levels[1:11], "Winkler I", "Winkler J", "Production Run BSC", "Full Enrichment BSC", "SEBPR", "FEBPR", "GEBPR","NF", "MWWTP", "scalemodel")
sample_data(phyobj_all)$System.Name = levels(corlevels)


# Facet wrap 
top_nested <- nested_top_taxa(phyobj_all,
                          top_tax_level = "Class",
                          nested_tax_level = "Order",
                          n_top_taxa = 10, 
                          n_nested_taxa = 1)
p = plot_nested_bar(top_nested$ps_obj,
                top_level = "Class",
                nested_level = "Order",
                legend_title = "Class and Order") +
  facet_wrap(~Bin,
             scales = "free_x",
             ncol = 6) +
  labs(title = "Relative abundances of the top 3 orders for each of the top 3 classes") +
  theme(plot.title = element_text(hjust = 0.5, 
                                  size = 20, 
                                  face = "bold"),
        legend.key.size = unit(10, 
                               "points"))
p
ggsave("top_3_classes_top5_order_bin.png", plot = p, path = working_directory, width = 8000, height = 4000, units = "px", dpi = 380)
```

```{r Top taxa poster figure generation}

# displays poorly in R markdown but saved image has correct dimensions #

#https://github.com/gmteunisse/Fantaxtic

library(extrafont)
library(showtext)
font_add(family = "skeena", regular = "\\\\wsl.localhost/Ubuntu/home/ruben/bioinfo/sludge/skeena_d.ttf")
showtext_auto()

#Find top 10 across all samples
norm_ps = merge_samples(phyobj_all, "Bin")
norm_ps@sam_data$Bin = factor(norm_ps@sam_data$Bin, levels = c("PN", "P", "PHA"))
top_asv <- top_taxa(norm_ps, n_taxa = 10, tax_level = "Order")

# options(repr.plot.width = 4, repr.plot.height =1.5) 
p3 = plot_nested_bar(ps_obj = top_asv$ps_obj,
                     top_level = "Class",
                     nested_level = "Order",
                     palette = c(Alphaproteobacteria = "#F3722C", 
                                 Bacteroidia = "#871762",
                                 Gammaproteobacteria = "#FFBB33"),
                     merged_clr = "#919191",
                     sample_order = c("PN", "P", "PHA"),
                     legend_title = "Orders by Class") +
  labs(title = "10 most represented orders") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black", family = "skeena", size = 40),
        axis.text.y = element_text(family = "skeena", size = 40),
        title = element_text(family = "skeena", size = 50),
        plot.background = element_rect(fill = "#F0EEE4"),
        plot.title = element_text(hjust = 0.5), 
        legend.background = element_blank(),
        legend.key = element_blank(),
        legend.title = element_text(family = "skeena", size = 40),
        legend.text = element_text(size = 40, family = "skeena")) +
  xlab("")
  


p3
ggsave("top_10_orders_bin.png", plot = p3, path = working_directory, width = 7, height = 4.5, units = "in", dpi = 300)
```

```{r ANCOMBC analysis}
# Code courtesy of Dr. Maxine Wren, modified for this data

# remove collinear variables from the metadata
filtered_meta = meta %>% dplyr::select(-c("System_Type", "Media_Type_1"))
ASV = otu_table(backup_phyobj_all)
TAX = tax_table(backup_phyobj_all)
SAMPLE = sample_data(filtered_meta)
filtered_ps = phyloseq(ASV, TAX, SAMPLE)

## remove any taxa with <10 reads in the total data set (sum of all samples <10). This is because extremely rare taxa can produce a series of false positives. 

filtered_ps = prune_taxa(taxa_sums(filtered_ps) > 10, filtered_ps) 

#check the results of filtering
otu_df = otu_table(filtered_ps) %>% as.data.frame() %>% mutate(sums = rowSums(.)) %>% dplyr::select(sums, everything())

# Continuous covariate: "Collection.Date"
# Categorical covariates: “System.Name, Bin”
# The group variable of interest: “Bin”
# Three groups: "P removal", "N and P removal", 
# The reference group: “Full scale WRRF”

# refactor the categorical variables so PHA bioreactor is the comparator
sample_data(filtered_ps)$Bin <- factor(sample_data(filtered_ps)$Bin,
                                      levels=c("PHA bioplastics reactor", 
                                               "N and P removal", 
                                               "P removal"))
# If I do the fix_formula as System.Name + Collection.Date + Bin it gives me an error about checking for collinearity or missing values in my covariates, so I am just running it with "bin" as the formula.

ancombc_out = ancombc2(filtered_ps,
                             fix_formula = "Bin",
                             tax_level = "Order",
                             group = "Bin",
                             struc_zero = TRUE,
                             neg_lb = TRUE,
                             verbose = TRUE,
                             alpha = 0.05,
                             global = TRUE,
                             pairwise = TRUE,
                             dunnet = TRUE
                             )
```

```{r taxon presence/absence}
# taxon presence/absence
tab_zero = ancombc_out$zero_ind %>% column_to_rownames(var = "taxon") #%>% mutate(sum = rowSums(.)) %>% filter(sum == 2)

res_prim = ancombc_out$res
res_global= ancombc_out$res_global %>% mutate_if(is.numeric, function(x) round(x, 5))
res_global_passed = res_global %>% filter(passed_ss == TRUE) %>% filter(diff_abn == TRUE)
diff_taxons = res_global_passed$taxon
diff_taxons
```



# Bray Curtis NMDS 

```{r variance stabilized NMDS: Bin whole sample}
set.seed(100)
#Bray-Curtis distance matrix calculation by System Type

physeq_ord <- ordinate(
 physeq = phyobj_all,
 method = "NMDS",
 formula = ~Bin,
 distance = "bray"
)
stress = as.character(round(physeq_ord$stress, digits = 3))
stresslab = paste("Stress =", stress)

p1 = plot_ordination(
 physeq = phyobj_all,
 ordination = physeq_ord,
 color = "Bin"
) +
 aes(group = Bin) +
 geom_point(aes(color = Bin), alpha = 0.95, size = 6) +
  scale_color_manual(values=c("#A6CEE3","#347F92","#94DE11"))+
 stat_ellipse(type = "t", geom = "polygon", position = "identity", alpha = .05, aes(fill = Bin)) + 
 scale_fill_manual(values=c("#A6CEE3","#347F92","#94DE11")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
    panel.background = element_blank(), axis.line = element_line(colour = "black"),strip.text = element_text(colour = 'white'),text = element_text(size=12),legend.text=element_text(size=12),legend.title = element_blank())+
  annotate("text", x = 2, y = -2, label = stresslab) + 
  annotate("text", x = 2, y = -2.5, label = "Non-metric fit, R² = 0.985")

p1
ggsave("nmds_Bin.png", plot = p1, path=working_directory)

#filter out ASVs that have only 0 values
ot_df = otu_table(phyobj_all) %>% as.data.frame()

#247 -> 239
ot_df = ot_df %>% mutate(sum = rowSums(.)) %>% dplyr::select(sum, everything()) %>% filter(sum != 0) %>% dplyr::select(-sum) %>% t()
samples = rownames(ot_df)
adonis_dist = vegdist(ot_df, method = "bray") 

# adonis_mds = metaMDS(adonis_dist) %>% scores() %>% 
#   as_tibble(rownames="Group")

#subset samples needed from meta
# needed_samples = rownames(ot_df)
# parcu_adonis_meta = meta %>% rownames_to_column(var="Group") %>% filter(Group %in% needed_samples)
# 
# meta_nmds = inner_join(adonis_mds, parcu_adonis_meta, by = "Group")

# centroid_systemtype = meta_nmds %>% 
#   group_by(Bin) %>% 
#   summarize(NMDS1 = mean(NMDS1), NMDS2 = mean(NMDS2))
# 
# meta_nmds %>% 
#   ggplot(aes(x=NMDS1, y = NMDS2, color = Bin))+
#   geom_point() +
#   stat_ellipse(show.legend = FALSE) +
#   geom_point(data = centroid_systemtype, size = 2, shape = 21, color = "black", aes(fill=Bin), show.legend = FALSE)+
#  # ggtitle("Media Type, bray-curtis")+
#   scale_colour_discrete(name="Bin") +
#   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
#     panel.background = element_blank(), axis.line = element_line(colour = "black"))

stressplot(physeq_ord)

submeta = meta %>% t() %>% as.data.frame %>% dplyr::select(any_of(samples)) %>% t() %>% as.data.frame()

# PERMANOVA # 
bintype = adonis2(adonis_dist ~ Bin, data = submeta, permutations = 1e4)
bintype$`Pr(>F)`[1] #p-value 

bd = betadisper(adonis_dist, submeta$Bin)
anova(bd)


```

```{r NMDS by bin, for poster}
# displays poorly in R but saves image with correct dimensions # 

physeq_ord <- ordinate(
 physeq = phyobj_all,
 method = "NMDS",
 formula = ~Bin,
 distance = "bray"
)

stress = as.character(round(physeq_ord$stress, digits = 3))
stresslab = paste("Stress =", stress)

p1 = plot_ordination(
 physeq = phyobj_all,
 ordination = physeq_ord,
 color = "Bin"
) +
 aes(group = Bin) +
 geom_point(aes(color = Bin), alpha = 0.95, size = 5) +
  scale_color_manual(values=c("#F3722C","#871762","#90B30E"))+
  scale_fill_manual(values=c("#F3722C","#871762","#90B30E")) +
  stat_ellipse(type = "t", geom = "polygon", position = "identity", alpha = .05, aes(fill = Bin)) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
    panel.background = element_blank(), axis.line = element_line(colour = "black"),strip.text = element_text(colour = 'white'),text = element_text(size=60, family = "skeena"),legend.text=element_text(size=60),legend.title = element_blank())+
  annotate("text", x = 4.2, y = -2, label = stresslab, size = 12, family = "skeena")+
  annotate("text", x = 3.6, y = -2.5, label = "Non-metric fit, R² = 0.985", size = 12, family = "skeena")+
  theme(plot.background = element_rect(fill = "#F0EEE4"), legend.background = element_rect(fill = "#F0EEE4"))
  

p1
ggsave("nmds_Bin.png", plot = p1, path=working_directory, height = 5, width = 7, dpi = 300)
```

Efforts were made at the beginning of the project to investigate Parcubacteria specifically. Beta diversity analysis using NMDS proved difficult due to the sparsity of Parcubacteria differences between samples. Presence-absence and abundance testing was possible, however, and revealed that PN systems and systems with real wastewater had a greater amount and variety of Parcubacteria.