---
title: "DADA2_sludge_2023"
output: html_document
date: "2023-10-23"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("dada2")
library("devtools")
library("phyloseq")
library("Biostrings")
library("ggplot2")
library("gridExtra")
library("dplyr")
library(tidyverse)
library(Biostrings); packageVersion("Biostrings")
library(DECIPHER)
library(phangorn)
library(phyloseq)
library(tibble)
library(ggh4x)
library(microbiome)
library(viridis)

```



```{r Set up directory}
path_R1 <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/R1_raw_reads" # CHANGE ME to the directory containing the fastq files after unzipping.
path_R2 <- "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/R2_raw_reads" # CHANGE ME to the directory containing the fastq files after unzipping.
```

```{r Extract sample names}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1.fastq and SAMPLENAME_R2.fastq
fnFs <- sort(list.files(path_R1, pattern="_R1.fastq", full.names = TRUE))
fnRs <- sort(list.files(path_R2, pattern="_R2.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```


```{r Inspect read quality profiles}
# Create the plots
plot1 <- plotQualityProfile(fnFs[1:2])
plot2 <- plotQualityProfile(fnRs[1:2])

# Combine plots into a single figure
grid.arrange(plot1, plot2, nrow = 1)

```

```{r Filter and trim}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path_R1, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path_R2, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(260,240), 
              maxN=0, maxEE=c(3,6), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```


```{r Learn the Error Rates}

errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

plotErrors(errF, nominalQ=TRUE)

```

```{r Sample Inference}
dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaRs <- dada(filtRs, err=errR, multithread=TRUE)
dadaFs[[1]]
```


```{r Merge paired reads}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=FALSE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])

library(ShortRead) ## Load this package first for this code to work consistently
library(dada2)
merged.seqs <- mergers$str_extract(EUB)["sequence"]  # Or wherever the list of seqs to output is
names(merged.seqs) <- paste0("ASV", seq_along(merged.seqs)) # e.g.
writeFasta(merged.seqs, "merger_dada2.fasta")
```


```{r Construct amplicon sequence variant (ASV) table}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

```

```{r Remove chimeras}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

sum(seqtab.nochim)/sum(seqtab)
```


```{r Track reads through the pipeline}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```


```{r Assign taxonomy}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/silva_nr99_v138.1_train_set.fa", multithread=TRUE)

taxa2 <- addSpecies(taxa, "/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/silva_species_assignment_v138_1.fa")
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
``` 

```{r Transpose asv df to have ASVs as the rownames}
asv <- t(seqtab.nochim)
```

```{r Setting up a dataframe for phyloseq}

library(readxl)
metadata <- as.data.frame(read_excel("/Users/leycufur/bioinfo/Bi624/Sludge/Illumina_Reads_updated_1-10-24.xlsx"))
colnames(metadata) <- gsub(" ", "_", colnames(metadata)) #replacing spaces in column names with underscores

# converting the Sample_ID column to the row names
meta <- metadata %>% column_to_rownames("Sample_ID")

# adding the sample_ID back into the df
meta$Sample_ID <- rownames(meta)

```


```{r Construct phyloseq object}
asv = as.data.frame(asv) %>% mutate(count = rowSums(.)) %>% dplyr::select(count, everything()) %>% filter(count > 1) %>% dplyr::select(-count)

ps <- phyloseq(otu_table(asv, taxa_are_rows = TRUE), 
               sample_data(meta), 
               tax_table(taxa))

```


```{r Renaming ASVs to short names}
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))

```

```{r Creating a phyloseq object with a subset of only parcubacteria }
parcu_ps <- subset_taxa(ps, Class=="Parcubacteria")
```


```{r Plot richness}
plot_richness(ps,x = "Media_Type_1", measures=c("Shannon", "Simpson"), color="Media_Type_1")
plot_richness(parcu_ps,x = "Media_Type_1", measures=c("Shannon"))

```

```{r from Max}

asv_df = as.data.frame(asv)

asv_names = tibble::rownames_to_column(asv_df, "ASV")

long_asv = pivot_longer(asv_names,
                        cols = 2:89,
                        names_to = "Group")
long_asv = rename(long_asv, "name" = "ASV")

shared = long_asv
```

```{r Writing a function for alpha diversity (From Max)}
richness <- function(x){sum(x > 0)}
shannon <- function(x){
  relativeabund <- x[x>0]/sum(x)
  -sum(relativeabund * log(relativeabund))
}
simpson <- function(x){
  n <- sum(x)
  sum(x*(x-1) / (n*(n-1)))
}

summary_shared = shared %>% 
  group_by(Group) %>% 
  summarize(richness = richness(value),
            shannon = shannon(value),
            simpson = simpson(value),
            invsimpson = 1/simpson(value),
            n = sum(value))

summary_shared %>% 
  pivot_longer(cols = c(richness, shannon, simpson, invsimpson),
               names_to = "metric") %>% 
  ggplot(aes(x=n, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 4, scales = "free_y")
```

```{r preprocessing, variance stabilization (normalization)}
# correct for different read depths and variance stabilizing
#Code from Dr Max Wren
library(DESeq2)
#First, we need to create a DESeq2 object. Selected experimental variable is Bin (expected function)
diagdds <- phyloseq_to_deseq2(ps, ~Bin)
#Next, we need to estimate the variance between samples
diagdds <- estimateSizeFactors(diagdds,type="poscounts")
diagdds <- estimateDispersions(diagdds)
#Now we need to use the variance between samples to normalize the data.
#This reduces "noise" from variation between samples that may be due to differences
#in amplification or sequencing depth.
diagvst <- getVarianceStabilizedData(diagdds)
#Now we need to make it BACK into an ASV/OTU table and slot it back into the phyloseq.
otu_table(ps) <- otu_table(diagvst, taxa_are_rows = TRUE)
ot_df <- data.frame(otu_table(ps))
ot_df[ot_df < 0.0] <- 0.0 # why this?
ot <- otu_table(ot_df, taxa_are_rows = TRUE)
sample_names(ot) <- sample_names(ps@otu_table)
otu_table(ps) <- otu_table(ot, taxa_are_rows = TRUE)
```


```{r NMDS plots by bin}

 myColors <- c('#AACC00', '#D4D700', '#7D451B', '#007F00', '#80B918', '#55A630', '#583101', '#A08042', '#AFFC41', '#A4B75C','#F8961E', '#F3722C', '#FFBB33', '#CCAA22','#D00000', '#9A040F','#647332','#9D4EDD', '#E0AAFF', '#AB51E3', '#3C096C')

nmdsColors <- c("#7D451B", "#F8961E", "#007F00")
set.seed(777)



samples <- sample_sums(ps) > 10000
ps.prop <- prune_samples(samples, ps)

# Transform data to proportions as appropriate for Bray-Curtis distances
ps.prop <- transform_sample_counts(ps, function(otu) otu/sum(otu)) 
ps.prop <- subset_taxa(ps, Kingdom == "Bacteria")
ps.prop <- subset_taxa(ps.prop, Family != "Mitochondria")
ps.prop <- subset_taxa(ps.prop, Order != "Chloroplast")


ord.nmds.bray <- ordinate(ps.prop, method = "NMDS", distance = "bray")

# Extract the stress value
stress_value <- round(ord.nmds.bray$stress, digits = 3)

# Plot NMDS with stress and R-squared values on the plot
p3 <- plot_ordination(ps.prop, ord.nmds.bray, color = "Bin") + 
  scale_color_manual(values = nmdsColors) + 
  theme(
    plot.background = element_rect(fill = "#F0EEE4"),
    legend.background = element_rect(fill = "#F0EEE4"),
    panel.background = element_rect(fill = "white", colour = "white"),
    panel.grid.major = element_line(linewidth = 0.5, linetype = 'solid',
                                colour = "gray"), 
  ) + 
  labs(fill = NULL) + 
  geom_text(aes(x = 1, y = 0, label = paste("Stress:", stress_value)), color = "black", size = 3, hjust = -1, vjust = 11) + 
  geom_text(aes(x = 1, y = 0, label = paste("R-squared: 0.983")), color = "black", size = 3, hjust = -.76, vjust = 12.3)

p3
ggsave("NMDS.png", p3, dpi = 600)





```

```{r Filter phyloseq and plot nmds with bins}
 

set.seed(777)

ps.prop = subset_taxa(ps.prop, Kingdom == "Bacteria")
ps.prop = subset_taxa(ps.prop, Family != "Mitochondria")
ps.prop = subset_taxa(ps.prop, Order != "Chloroplast")
ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance="bray")

ps.prop
p = plot_ordination(ps.prop, ord.nmds.bray, color="Bin", title="Bray NMDS")

p + scale_color_manual(values = myColors)

```



```{r Abundance}
theme_set(theme_bw())
gp.ch = subset_taxa(ps, Class == "Parcubacteria")
p <- plot_bar(gp.ch, x = "Media_Type_1", fill = "Class")
p + geom_bar(aes(color=Order, fill=Order), stat="identity", position="stack")

```

```{r Parcu relative abundance}

phy <- subset_taxa(ps, Class == "Parcubacteria") # Subset for parcubacteria only
get_taxa_unique(phy, "Order")
phy2 <- phy %>% aggregate_taxa(level = "Order") %>%
  microbiome::transform(transform = "compositional")
zero_abundance_samples <- sample_sums(otu_table(phy2)) == 0 #filtering out samples with no data for abundance
phy2_filtered <- prune_samples(!zero_abundance_samples, phy2) # Remove samples with zero abundances from phy2

head(otu_table(phy2_filtered))
head(tax_table(phy2_filtered))

# Plot the compositional abundance by Media_Type_1 after filtering
mediatype_parcu_relabund <- phy2_filtered %>% plot_composition(group_by = "Media_Type_1") +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_viridis(option = "B", discrete = TRUE)
mediatype_parcu_relabund + force_panelsizes(total_width = unit(20, "inch"))

# Plot the compositional abundance by System_Type after filtering
systype_parcu_relabund <- phy2_filtered %>% plot_composition(group_by = "System_Type") +
  scale_y_continuous(labels = scales::percent) +
  scale_fill_viridis(option = "D", discrete = TRUE)
systype_parcu_relabund + force_panelsizes(total_width = unit(20, "inch"))



# phy2 %>%  plot_composition(group_by = "Media_Type_1")+
#   scale_y_continuous(labels = scales::percent) +
#   scale_fill_viridis(discrete = TRUE)

```

```{r Whole data bacterial relative abundance}
phy3 <- ps %>% aggregate_taxa(level = "Class") %>%
          microbiome::transform(transform = "compositional")

mediatype_whole_relabund <- phy3 %>%  plot_composition(group_by = "Media_Type_1")+
  scale_y_continuous(labels = scales::percent)+
  scale_fill_viridis(option = "B", discrete = TRUE)
mediatype_whole_relabund + force_panelsizes(total_width = unit(20, "inch"))

systype_whole_relabund <- phy3 %>%  plot_composition(group_by = "System_Type")+
  scale_y_continuous(labels = scales::percent)+
  scale_fill_viridis(option = "D", discrete = TRUE)
systype_whole_relabund + force_panelsizes(total_width = unit(20, "inch"))

```

```{r I'm still trying to figure out what I was doing here}
    library(phyloseq)
    library(ggplot2)
    library(magrittr)

    # To simplify the GP data
    ps = subset_taxa(ps, Kingdom == "Bacteria")
    ps = prune_taxa(taxa_sums(ps) > 1000, ps)

    total = median(sample_sums(ps))
    standf = function(x, t=total) round(t * (x / sum(x)))
    normalized_ps = transform_sample_counts(ps, standf)

    # Merge sample by their sampling site
    # Here we have 9 "samples" (SampleType)
    merged_ps_phylum = merge_samples(normalized_ps, "Media_Type_1") # summed
    merged_ps_phylum
    #> phyloseq-class experiment-level object
    #> otu_table()   OTU Table:         [ 1973 taxa and 9 samples ]
    #> sample_data() Sample Data:       [ 9 samples by 7 sample variables ]
    #> tax_table()   Taxonomy Table:    [ 1973 taxa by 7 taxonomic ranks ]
    #> phy_tree()    Phylogenetic Tree: [ 1973 tips and 1972 internal nodes ]

    # Sample Data has Sample ID per row
    normalized_ps_phylum1 = tax_glom(normalized_ps, "Class", NArm = TRUE)
    normalized_ps_phylum1
    #> phyloseq-class experiment-level object
    #> otu_table()   OTU Table:         [ 26 taxa and 26 samples ]
    #> sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
    #> tax_table()   Taxonomy Table:    [ 26 taxa by 7 taxonomic ranks ]
    #> phy_tree()    Phylogenetic Tree: [ 26 tips and 25 internal nodes ]

    # Sample Data has Sample Type per row
    normalized_ps_phylum2 = tax_glom(merged_ps_phylum, "Class", NArm = TRUE)
    normalized_ps_phylum2
    #> phyloseq-class experiment-level object
    #> otu_table()   OTU Table:         [ 26 taxa and 9 samples ]
    #> sample_data() Sample Data:       [ 9 samples by 7 sample variables ]
    #> tax_table()   Taxonomy Table:    [ 26 taxa by 7 taxonomic ranks ]
    #> phy_tree()    Phylogenetic Tree: [ 26 tips and 25 internal nodes ]

    normalized_ps_phylum1_relabun = transform_sample_counts(normalized_ps_phylum1, function(OTU) OTU/sum(OTU) * 100)
    normalized_ps_phylum2_relabun = transform_sample_counts(normalized_ps_phylum2, function(OTU) OTU/sum(OTU) * 100)

    psmelt(normalized_ps_phylum1_relabun) %>% #phylum relative abundance by sammple
            ggplot(aes(x = Sample, y = Abundance, fill = Phylum)) +
            geom_bar(stat = "identity") +
            theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
            labs(y = "Relative Abundance", title = "Phylum Relative Abundance")
    
    
    psmelt(normalized_ps_phylum2_relabun) %>% #phylum relative abundance by wwt
            ggplot(aes(x = Sample, y = Abundance, fill = Phylum)) +
            geom_bar(stat = "identity") +
            theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
            labs(y = "Relative Abundance", title = "Phylum Relative Abundance")
    
    
# Filter for the top 10 classes
top_classes <- names(sort(taxa_sums(normalized_ps_phylum2_relabun), decreasing = TRUE))[1:16]
normalized_ps_phylum2_relabun_top10 <- prune_taxa(top_classes, normalized_ps_phylum2_relabun)

# Create a bar plot for the top 10 classes
psmelt(normalized_ps_phylum2_relabun_top10) %>% #phylum relative abundance by wwt for top 10 phylum
  ggplot(aes(x = Sample, y = Abundance, fill = Phylum)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(y = "Relative Abundance", title = "Top 10 Phylum Relative Abundance")

# Filter for the top 10 classes based on total abundance across samples
top_classes <- names(sort(sample_sums(normalized_ps_phylum2_relabun), decreasing = TRUE))[1:10]
normalized_ps_phylum2_relabun_top10 <- prune_taxa(top_classes, normalized_ps_phylum2_relabun)

# Create a bar plot for the top 10 classes
psmelt(normalized_ps_phylum2_relabun_top10) %>%
  ggplot(aes(x = Sample, y = Abundance, fill = Phylum)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(y = "Relative Abundance", title = "Top 10 Phylum Relative Abundance")


```

```{r Top 10 classes by sample type}
    library(phyloseq)
    library(tidyverse)

    # To simplify the GP data
    ps = subset_taxa(ps, Kingdom == "Bacteria")
    ps = prune_taxa(taxa_sums(ps) > 1000, ps)

    # Merge sample by their sampling site
    merged_ps = merge_samples(ps, "Media_Type_1") # summed

    # Merge taxa by 'Class rank
    normalized_ps_phylum = tax_glom(merged_ps, "Class", NArm = TRUE)

    # Calculate relative abundance
    normalized_ps_phylum_relabun = transform_sample_counts(normalized_ps_phylum, function(OTU) OTU/sum(OTU))

    # Show taxa_sums
    taxaSums = data.frame(tax_table(normalized_ps_phylum_relabun)[,"Class"],
                          taxa_sums = taxa_sums(normalized_ps_phylum_relabun)) %>%
            arrange(desc(taxa_sums)) # reverse sort
    taxaSums
    # Select top10 phyla
    top10 = head(rownames(taxaSums), 10)
    top10

    # Remove unselected phyla
    y = prune_taxa(top10, normalized_ps_phylum_relabun)
    y

        # Build data.frame
    df1 = data.frame(ID = c(taxa_names(y)), Class = c(tax_table(y)[,"Class"]))
    df1

    df2 = t(cbind(otu_table(y)))
    df2


    df = cbind(df1, df2) %>%
            pivot_longer(-c(ID, Class), names_to = "Media_Type_1", values_to = "Abundance") %>%
            as.data.frame
    head(df)

        # Re-level Phylum by abundance
    df$Class = as.factor(df$Class)
    levels(df$Class) # Before ordering

        df$Phylum = factor(df$Class, levels = c(taxaSums[top10,"Class"]))
    levels(df$Phylum) # After ordering

    

ggplot(df, aes(Media_Type_1, Abundance, fill = Class)) +
  geom_col(color = "black") +
  scale_fill_manual(values = myColors) +  # Use scale_fill_manual for custom colors
  scale_y_continuous(labels = scales::percent_format(accuracy = 1)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  labs(x = "Sample type", y = "Average relative abundance")

```




```{r Rarefaction}

rand <- shared %>% 
  uncount(value) %>% 
  mutate(name = sample(name)) %>% 
  count(Group, name, name = "value")

summary_rand = rand %>% 
  group_by(Group) %>% 
  summarize(richness = richness(value),
            shannon = shannon(value),
            simpson = simpson(value),
            invsimpson = 1/simpson(value),
            n = sum(value))

summary_rand %>% 
  pivot_longer(cols = c(richness, shannon, simpson, invsimpson),
               names_to = "metric") %>% 
  ggplot(aes(x=n, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 4, scales = "free_y")+
  ggtitle("Random sampling")

```

```{r }
library(vegan)
summary_shared = shared %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            invsimpson = 1/diversity(value, index = "simpson"),
            n = sum(value))

summary_shared %>% 
  pivot_longer(cols = c(richness, shannon, simpson, invsimpson),
               names_to = "metric") %>% 
  ggplot(aes(x=n, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 4, scales = "free_y")

```


```{r This is from Max I don't think I ended up using any of this}
# 
# parcubacteria_only = cbind(as.data.frame(taxa), as.data.frame(asv)) %>% filter(Class == "Parcubacteria") %>% select(-c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))
# 
# parcubacteria_only = t(parcubacteria_only)
# parcubacteria_only = as.data.frame(parcubacteria_only)
# 
# parcubacteria_only = parcubacteria_only %>% mutate(num_asvs = rowSums(.)) %>% select(num_asvs, everything())
# 
# #using a threshold of 10,000
# filtered_tasv = parcubacteria_only %>% filter(num_asvs > 1) %>% select(-num_asvs)
# 
# long_filt_tasv = filtered_tasv %>% rownames_to_column("Group") %>% pivot_longer(-Group)
# 
# #remove ASVs with a total of 0 counts after removing ones below the threshold
# long_filt_tasv = long_filt_tasv %>% group_by(name) %>% mutate(total = sum(value)) %>% filter(total != 0)
# 
# working_asv = long_filt_tasv %>% ungroup() %>% select (-total) %>% pivot_wider(id_cols = Group) %>% as.data.frame() %>% column_to_rownames("Group") %>% as.matrix()
# 
# #use vegdist to calculate dist first 
# 
# braydist = vegdist(working_asv, method = "bray")
# 
# set.seed(100)
# nmds = metaMDS(braydist)
# 
# scores = scores(nmds)
# 
# scores(nmds) %>% as_tibble(rownames = "Group") %>% 
#   ggplot(aes(x = NMDS1, y = NMDS2)) +
#   geom_point()

```

```{r }
data(enterotype)

ps_subset = subset_samples(ps, !is.na(Media_Type_1))

```


```{r Did not use Microbiome Network Plot using ggplot2}
plot_net(ps, maxdist = .8, point_label = "Sample_ID", color = "Media_Type_1")
```

```{r Did not useMicrobiome Network Plot using ggplot2 but rainbow}
ig <- make_network(ps, max.dist = 0.8, dist.fun = "bray")
plot_network(ig, ps, color = "System_Name", shape = "Media_Type_1")
```

```{r Did not use Run alignment with DECIPHER}
# extract sequences from DADA2 output
sequences<-getSequences(seqtab.nochim)
names(sequences)<-sequences

# Run alignment
alignment <- AlignSeqs(DNAStringSet(sequences), anchor=NA)

# Alignment takes a lot of time and introduces a lot of gaps. Ended up not using this alignment or the file it made

```

```{r Did not use Export alignment to fasta}
# Assuming 'alignment' is the alignment object obtained from DECIPHER

# Convert alignment object to a character matrix
alignment_matrix <- as(alignment, "matrix")

# Write the character matrix to a FASTA file
library(seqinr)  # Load the seqinr package for write.fasta function

# Replace 'alignment_file.fasta' with your desired file name
write.fasta(sequences = alignment_matrix, names = colnames(alignment_matrix), file.out = "alignment_file.fasta")

```


```{r Change sequence alignment output into a phyDat structure}
library(phangorn)

phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) # Note, tip order != sequence order


# negative edges length changed to 0!
fit = pml(treeNJ, data=phang.align)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                    rearrangement = "stochastic", control = pml.control(trace = 0))
```

```{r Adding phylogenetic info into phyloseq object}
ps_phy_tree <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               tax_table(taxa),phy_tree(fitGTR$tree))

```

```{r changing OTU names into shortened names}
taxa_names_og <- taxa_names(ps_phy_tree)
taxa_names(ps_phy_tree) <- 1:ntaxa(ps_phy_tree)
taxa_names(ps_phy_tree) <- paste("OTU-", taxa_names(ps_phy_tree), sep="")
taxa_names(ps_phy_tree)

```

```{r creating phylogenetic tree}
phy = phy_tree(ps_phy_tree)
treeSegs0 = tree_layout(phy)
treeSegs1 = tree_layout(ps_phy_tree)
edgeMap = aes(x=xleft, xend=xright, y=y, yend=y)
vertMap = aes(x=x, xend=x, y=vmin, yend=vmax)
p0 = ggplot(treeSegs0$edgeDT, edgeMap) + geom_segment() + geom_segment(vertMap, data=treeSegs0$vertDT)
p1 = ggplot(treeSegs1$edgeDT, edgeMap) + geom_segment() + geom_segment(vertMap, data=treeSegs1$vertDT)
print(p0)
print(p1)
plot_tree(ps_phy_tree, "treeonly")
plot_tree(ps_phy_tree, "treeonly", ladderize="right")

phylo_tree <- plot_tree(ps_phy_tree, "treeonly", ladderize="right")
phylo_tree

dev.new()
phylo_tree

pdf("phylo_tree.pdf", width = 200, height = 45)
phylo_tree
dev.off()
```


```{r Export OTU to tsv}
write_biom_csv <- function(ps_phy_tree, file, sep = "; ") {
  phyloseq::otu_table(ps_phy_tree) %>%
    as.data.frame() %>%
    rownames_to_column("#OTU ID") %>%
    left_join(phyloseq::tax_table(ps_phy_tree) %>% 
                as.data.frame() %>%
                rownames_to_column("#OTU ID") %>% 
                tidyr::unite("taxonomy", !`#OTU ID`, sep = sep)) -> phyloseq_biom
  
  write_tsv(phyloseq_biom, file = file)
}

write_biom_csv(ps_phy_tree, "phyloseq_biom.tsv")
```

```{r generating multiple sequence alignment file}
writeXStringSet(alignment,
   file="/Users/leycufur/bioinfo/Bi624/Sludge/DADA2/msa.fasta")
```

```{r Export ASV table as tsv}

# Convert ASV table to a data frame
asv_table <- as.data.frame(asv_names)  # Convert OTU table to a data frame

# Write ASV table to a CSV file
write_tsv(asv_table, file = "asv_table.tsv")
```

```{r}

top10 <- names(sort(taxa_sums(ps), decreasing=TRUE))[1:10]
ps.top20 <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.top20 <- prune_taxa(top20, ps.top20)
plot_bar(ps.top20, x="Media_Type_1", fill="Class") 

```

```{r}

```

