---
title: "FAPROTAX Figures"
author: "Farris Tedder, some code adapted from Ruben Lancaster"
date: 3-6-2024
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, message = F, warning = F)
library(tidyverse)
library(dplyr)
library(ggplot2)
library("readxl")
library(data.table)
```

```{r Read in asv / tax files} 
#Files are generated by FAPROTAX

#The asv file can be formatted either by frequency or by counts.
#asv <- read_tsv("C:/Users/farri/Coding/Sludge/asv_table_leyla_1-9-23.tsv")
#fapro_raw = read_tsv("C:/Users/farri/Coding/Sludge/silva_allgroups_taxtable_1-25-24-filtered_out.tsv")
fapro_raw = read_tsv("files/silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv")

#Taxonomy assignment file
#tax <- read_csv("C:/Users/farri/Coding/Sludge/taxa_silva_dada2_20240110.csv")
report <- read_tsv("files/Funtional_Taxonomy_Report.txt")

#fix under counting NA in data file
total_na = report %>% filter(is.na(report$order))
report= mutate(report, order_total=replace(order_total, is.na(order), sum(total_na$count)))
metapath = "C:/Users/farri/Coding/Sludge/Sludge_Metadata_1_10_24.xlsx"

```

```{r Read in faprotax and format into different objects}
#This section hard codes in a lot of row and column numbers. If there are changes to the dataset this will break or give incorrect information, be sure to edit as needed.
fapro = fapro_raw
fapro$group_total <- rowSums(fapro[,2:85])
faproOver0 = filter(fapro, as.numeric(group_total)>= 1) #only keeping functions seen at least once
fapro300 = filter(fapro, as.numeric(group_total)>= 300) #only keeping functions seen over 300 times
fapro5000 = filter(fapro, as.numeric(group_total)>= 5000) #only keeping functions seen over 5000 times

# These functions are known to be associated with waste water treatment
wwt_fapro = filter(fapro, fapro$group %in% c("aerobic_chemoheterotrophy", "chemoheterotrophy", "denitrification", "nitrate_denitrification", "nitrate_reduction",
"nitrate_respiration",
"nitrite_denitrification",
"nitrite_respiration", "nitrogen_respiration",
"nitrous_oxide_denitrification", "Sample_ID"))

fapro_Long = transpose(fapro_raw)
rownames(fapro_Long) = colnames(fapro_raw)
colnames(fapro_Long) <- fapro_raw$group
fapro_Long = fapro_Long[-1,]
fapro_Long[, 1:92] <- sapply(fapro_Long[, 1:92], as.numeric)
fapro_Long$group_total <- rowSums(fapro_Long[,1:92]) #total number of functions per sample
fapro_Long = tibble::rownames_to_column(fapro_Long, "Sample_ID")

meta = read_excel(metapath)
names(meta)[names(meta) == "Bin"] <- "System_Focus" #More meaningful name
meta_temp = dcast(melt(meta, id.vars = "Sample_ID"), variable ~ Sample_ID)
meta_table <- meta_temp[,-1]
rownames(meta_table) <- meta_temp[,1]
meta_long = t(meta_table)
colnames(meta_temp)[1] = "group"
fullset = full_join(fapro_Long, meta, by="Sample_ID")
fullset[, 2:87] <- sapply(fullset[, 2:87], as.numeric)

#setting order for graphing
fullset$System_Type = as.character(fullset$System_Type)
fullset$System_Type = factor(fullset$System_Type, levels=c("Full scale WRRF", "Scale model WRRF", "Sequencing batch reactor"))

System_Order = c("Full scale WRRF", "Scale model WRRF", "Sequencing batch reactor")
Media_Order = c("Real wastewater", "Synthetic wastewater", "Fermented dairy manure")

# Setting functions known to be associated with waster water
wwt = fullset[,c("aerobic_chemoheterotrophy", "chemoheterotrophy", "denitrification", "nitrate_denitrification", "nitrate_reduction",
"nitrate_respiration",
"nitrite_denitrification",
"nitrite_respiration", "nitrogen_respiration",
"nitrous_oxide_denitrification", "Sample_ID")]
wwt = full_join(wwt, meta, by="Sample_ID")
wwt[, 1:10] <- sapply(wwt[, 1:10], as.numeric)
wwt$group_total <- rowSums(wwt[,1:10])#total number of functions per sample
tot = sum(wwt$group_total)
wwt = transform(wwt, group_percent = group_total/tot)

tot = sum(fullset$group_total)
fullset = transform(fullset, group_percent = group_total/tot)

# Colors used in graphing for consistancy
System_Colors = c("#347F92", "#A6CEE3", "#94DE11")
Media_Colors = c("#F58609", "#A01892", "#F2D000")
```

```{r general overview of system type data}
fullset %>%
  ggplot(aes(x=reorder(Sample_ID, -group_total), y=group_total, fill=System_Focus)) +
         geom_bar(stat="identity") + labs(y="Amount of reads tagged with functionality", x="Sample") + theme_light() + scale_fill_manual(values=c("#52ADC2", "#F58609","purple")) +theme( axis.text.y=element_text(size=25), axis.title=element_text(size=35), title=element_text(size=40), legend.text=element_text(size=25))+
  coord_flip()
#ggsave("Functionality_Totals.png", height=11, width=18, dpi=300)

wwt %>%
  ggplot(aes(x=reorder(Sample_ID, -group_total), y=group_total, fill=System_Focus)) +
         geom_bar(stat="identity")+ labs(y="Amount of reads tagged with WWT relevant functionality", x="Sample") + theme_light() + scale_fill_manual(values=c("#52ADC2", "#F58609","purple")) +theme( axis.text.y=element_text(size=25), axis.title=element_text(size=35), title=element_text(size=40), legend.text=element_text(size=25))+
  coord_flip()
#ggsave("Functionality_Totals_WWTTarget.png", height=11, width=18, dpi=300)

fullset %>%
  ggplot(aes(x=reorder(System_Focus, -group_total), y=group_total, fill=System_Focus)) +
         geom_bar(stat="identity")+ labs(y="Amount of reads tagged with functionality", x="Sample")+theme_light() + scale_fill_manual(values=c("#52ADC2", "#F58609","purple")) +theme( axis.text.y=element_text(size=25), axis.title=element_text(size=35), title=element_text(size=40), legend.text=element_text(size=25))+
  coord_flip()
#ggsave("Functionality_by_system_focus.png", height=11, width=18, dpi=300)

wwt %>%
  ggplot(aes(x=reorder(System_Focus, -group_total), y=group_total, fill=System_Focus)) +
         geom_bar(stat="identity")+ labs(y="Amount of reads tagged with wwt functionality", x="Sample") + theme_light() + scale_fill_manual(values=c("#52ADC2", "#F58609","purple")) + theme( axis.text.y=element_text(size=25), axis.title=element_text(size=35), title=element_text(size=40), legend.text=element_text(size=25))+
  coord_flip()
#ggsave("WWT_relivent_functionality_by_system_focus.png", height=11, width=18, dpi=300)

```

```{r relative functionality in top orders }
report$order_relative = report$count/report$order_total

report %>% filter(order %in% c("Bacteroidales", "Burkholderiales", "Chitinophagales", "Cytophagales", "Flavobacteriales", "Rhizobiales", "Rhodobacterales", "Rickettsiales")) %>%
  ggplot(aes(x=order, y=order_relative, fill=functionality)) +
         geom_bar(stat="identity")+ labs(y="Relative abundance of functionality", x="Order")+theme_light() +theme( axis.text.y=element_text(size=15), axis.title=element_text(size=20), title=element_text(size=40), legend.text=element_text(size=15))+
  coord_flip()
#ggsave("Functionality_relative_abundance_all_nofilter.png", height=11, width=18, dpi=300)

```

```{r relative functionality in top orders with data correction}
report_filtered = report[-c(2)] 
# The following functions were removed due to overlapping data:

  # [methanol_oxidation] "methanotrophy", "methylotrophy"               # [sulfur_respiration] "respiration_of_sulfur_compounds", "dark_oxidation_of_sulfur_compounds"      
  # [denitrificatoin] "nitrate_denitrification", "nitrite_denitrification", "nitrous_oxide_denitrification", "nitrate_reduction"                    
  # [human_associated] "human_pathogens_all", "human_gut", "mammal_gut", "aromatic_hydrocarbon_degradation"      
  # [nitrate_respiration and nitrogen_fixation] , "nitrogen_respiration",  "nitrite_respiration", "nitrification"
  # [photoautotrophy and photoheterotrophy] "anoxygenic_photoautotrophy_S_oxidizing", "oxygenic_photoautotrophy", "phototrophy", "photosynthetic_cyanobacteria"                        
# For a list of all functions removed run: setdiff(unique(report$functionality), unique(report_filtered$functionality))

# This removes the overlap of aerobic and regular chemeohetertroph so we can have both without overrepresentation
report_filtered = report_filtered %>% filter(functionality %in% c("chemoheterotrophy", "aerobic_chemoheterotrophy")) %>%
   filter(!duplicated(across(-functionality)))

report_filtered = rbind(report_filtered, report[-c(2)] %>% filter(functionality %in% c("denitrification", "	
sulfur_respiration", "thiosulfate_respiration", "iron_respiration", "nitrate_respiration", "dark_sulfide_oxidation", "aerobic_nitrite_oxidation", "aliphatic_non_methane_hydrocarbon_degradation", "animal_parasites_or_symbionts", "anoxygenic_photoautotrophy", "aromatic_compound_degradation", "methanol_oxidation", "cellulolysis", "chitinolysis", "chlorate_reducers", "chloroplasts", "dark_hydrogen_oxidation", "fermentation", "fumarate_respiration", "human_associated", "hydrocarbon_degradation", "intracellular_parasites", "manganese_oxidation", "nitrogen_fixation", "nonphotosynthetic_cyanobacteria", "photoautotrophy", "photoheterotrophy", "predatory_or_exoparasitic", "reductive_acetogenesis", "ureolysis")))

filtered_order_total = aggregate(count~order,report_filtered,sum)
names(filtered_order_total)[2] = 'filtered_order_total'
report_filtered = left_join(report_filtered, filtered_order_total, join_by(order))

report_filtered$order_relative_filtered = report_filtered$count/report_filtered$filtered_order_total

function_bar_order = c("aerobic chemoheterotrophy", "chemoheterotrophy", "fermentation", "denitrification",  "nitrate respiration", "nitrogen fixation", "cellulolysis", "ureolysis", "photoautotrophy", "photoheterotrophy", "animal parasites or symbionts", "chlorate reducers", "manganese oxidation", "hydrocarbon degradation", "aerobic nitrite oxidation", "dark sulfide oxidation", "dark hydrogen oxidation", "anoxygenic photoautotrophy", "iron respiration", "fumarate respiration", "human associated")

report_top_10 = report_filtered %>% filter(order %in% c("Bacteroidales", "Burkholderiales", "Chitinophagales", "Cytophagales", "Flavobacteriales", "Rhizobiales", "Rhodobacterales", "Sphingobacteriales", "Caulobacterales", "Sphingomonadales"))
report_top_10 = report_top_10 %>% mutate(functionality = gsub("_", " ", functionality))
report_top_10$functionality= factor(report_top_10$functionality, levels=function_bar_order)
```

```{r font}
library(extrafont)
library(showtext)
font_add(family = "Skeena Display", regular = "C:/Users/farri/AppData/Local/Microsoft/Windows/Fonts/skeena.ttf")
#font_import()
loadfonts(device = "win", quiet = TRUE)
showtext_auto()
```


```{r plot top 10 orders functionality}
report_top_10 %>% 
  ggplot(aes(x=order, y=order_relative_filtered, fill=functionality)) +
         geom_bar(stat="identity")+ labs(y="Relative abundance of functionality", x="Order")+ggtitle("Functionality in 10 most populous orders")+theme_light() +theme(text=element_text(family="Skeena Display"), plot.background = element_rect(fill = "#F0EEE4"), legend.background = element_rect(fill = "#F0EEE4"), axis.text.y=element_text(size=30), axis.title=element_text(size=35), plot.title = element_text(hjust = 0.5), title=element_text(size=40), legend.text=element_text(size=30),axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+coord_flip()+guides(fill=guide_legend(ncol =1))+
  scale_fill_manual(values = c('#90B30E', '#799900', '#AABA6E', '#D4D700', '#80B918', '#583101', '#7D3600', '#7D451B', '#826046', "#F3722C", '#F8961E','#ffbb33', '#D00000','#9A040F', '#C981B1', '#A35B8B','#7D3565', '#570F3F', '#4A4A4A','#666666', '#8F8F8F', '#B8B8B8','#D9D9D9'))

showtext::showtext_auto()
showtext::showtext_opts(dpi = 300)
#ggsave("Functionalities_Top_10_Orders.png", height=11, width=18, dpi=300)
```

```{r data processing: System_Focus by order}
fullset_filtered = fullset[!(row.names(fullset) %in% c("EUBF1","EUBB6","EUBB7", "EUBC7","EUBD1","EUBF5"  ,"EUBC9","EUBB12","EUBG9")),] 
fullset_filtered2 = data.frame(t(fullset_filtered))
colnames(fullset_filtered2) = fullset$Sample_ID

fullset_filtered = fullset_filtered2[-1,]
fullset_filtered = fullset_filtered[-(105:109),]
fullset_filtered = fullset_filtered[-(93:103),]
fullset_filtered2 = data.frame(t(fullset_filtered))
fullset_filtered = as.data.frame(sapply(fullset_filtered2, as.numeric))
fullset_filtered$System_Focus = fullset_filtered2$System_Focus

# Removing overlapping data. When non-overlapping regions can be conserved they are subtracted from parent function. Functions removed by name has no non-overlapping regions.
fullset_filtered$chemoheterotrophy = (fullset_filtered$chemoheterotrophy - fullset_filtered$aerobic_chemoheterotrophy)
fullset_filtered$phototrophy = (fullset_filtered$phototrophy - fullset_filtered$photoheterotrophy)
fullset_filtered$nitrate_reduction = (fullset_filtered$nitrate_reduction - fullset_filtered$nitrate_respiration)
fullset_filtered$photoautotrophy = (fullset_filtered$photoautotrophy - fullset_filtered$oxygenic_photoautotrophy)
fullset_filtered$methylotrophy = (fullset_filtered$methylotrophy - fullset_filtered$methanol_oxidation)
fullset_filtered$animal_parasites_or_symbionts = (fullset_filtered$animal_parasites_or_symbionts - fullset_filtered$human_associated)
fullset_filtered$hydrocarbon_degradation = (fullset_filtered$hydrocarbon_degradation - fullset_filtered$methanotrophy)
fullset_filtered$dark_oxidation_of_sulfur_compounds = (fullset_filtered$dark_oxidation_of_sulfur_compounds - fullset_filtered$dark_sulfide_oxidation)
fullset_filtered = fullset_filtered[ , -which(names(fullset_filtered) %in% c("nitrate_denitrification", "nitrite_denitrification", "nitrous_oxide_denitrification", "photosynthetic_cyanobacteria", "nitrogen_respiration", "human_gut", "mammal_gut", "human_pathogens_all", "aerobic_nitrite_oxidation"))]

System_Focus_PN = fullset_filtered %>% filter(System_Focus == "N and P removal")
System_Focus_PN = t(System_Focus_PN)
temp1 = data.frame(System_Focus_PN[1:83,1:26])
System_Focus_PN = as.data.frame(sapply(temp1, as.numeric))
rownames(System_Focus_PN) <- rownames(temp1)
System_Focus_PN$PN = rowSums(System_Focus_PN)
#System_Focus_PN=filter(System_Focus_PN, PN > 1500)
System_Focus_PN$functionality = rownames(System_Focus_PN)

System_Focus_P = fullset_filtered %>% filter(System_Focus == "P removal")
System_Focus_P = t(System_Focus_P)
temp2 = data.frame(System_Focus_P[1:83,1:26])
System_Focus_P = as.data.frame(sapply(temp2, as.numeric))
System_Focus_P$P = rowSums(System_Focus_P)
rownames(System_Focus_P) = rownames(temp2)
#System_Focus_P=filter(System_Focus_P, P > 1500)
System_Focus_P$functionality = rownames(System_Focus_P)

System_Focus_PHA = fullset_filtered %>% filter(System_Focus == "PHA bioplastics reactor")
System_Focus_PHA = t(System_Focus_PHA)
temp3 = data.frame(System_Focus_PHA[1:83,])
System_Focus_PHA = as.data.frame(sapply(temp3, as.numeric))
rownames(System_Focus_PHA) = rownames(temp3)
System_Focus_PHA$PHA = rowSums(System_Focus_PHA)
#System_Focus_PHA=filter(System_Focus_PHA, PHA > 1500)
System_Focus_PHA$functionality = rownames(System_Focus_PHA)

temp = data.frame(functionality=colnames(fullset_filtered))
np_tot=sum(System_Focus_PN$PN)
p_tot=sum(System_Focus_P$P)
pha_tot=sum(System_Focus_PHA$PHA)

System_Focus_PN=System_Focus_PN[, c("PN", "functionality")]
System_Focus_PN$PN = System_Focus_PN$PN/np_tot
System_Focus_P=System_Focus_P[, c("P", "functionality")]
System_Focus_P$P = System_Focus_P$P/p_tot
System_Focus_PHA=System_Focus_PHA[, c("PHA", "functionality")]
System_Focus_PHA$PHA = System_Focus_PHA$PHA/pha_tot

temp = merge(temp, System_Focus_PN, by="functionality", all=TRUE)
temp = merge(temp, System_Focus_P, by="functionality", all=TRUE)
temp = merge(temp, System_Focus_PHA, by="functionality", all=TRUE)
System_Focuss = pivot_longer(temp, cols = c("PN","P", "PHA"), names_to = "System_Focus", values_to = "count")
System_Focuss = na.omit(System_Focuss)
```

```{r relative abundance by system focus}
function_bar_order = c("aerobic chemoheterotrophy", "chemoheterotrophy", "fermentation", "denitrification",  "nitrate reduction","nitrate respiration", "nitrite respiration", "nitrogen fixation", "ureolysis", "photoheterotrophy",  "chlorate reducers", "manganese oxidation", "methanol oxidation", "dark sulfide oxidation", "dark hydrogen oxidation",  "aromatic compound degradation", "chloroplasts", "phototrophy", "oxygenic photoautotrophy", "nonphotosynthetic cyanobacteria", "human associated", "predatory or exoparasitic", "animal parasites or symbionts","intracellular parasites")
System_Focuss = System_Focuss %>% mutate(functionality = gsub("_", " ", functionality))
System_Focuss$functionality= factor(System_Focuss$functionality, levels=function_bar_order)

System_Focuss %>% mutate(System_Focus = fct_relevel(System_Focus, "PHA", "P","PN")) %>% ggplot(aes(x=System_Focus, y=count, fill=functionality)) +
  geom_bar(stat="identity")+ labs(y="Relative abundance of functionality", x="System")+ggtitle("Functionality in Systems")+theme_light() +theme(text=element_text(family="Skeena Display"), plot.background = element_rect(fill = "#F0EEE4"), legend.background = element_rect(fill = "#F0EEE4"), axis.text.y=element_text(size=30), axis.title=element_text(size=35), plot.title = element_text(hjust = 0.5), title=element_text(size=40), legend.text=element_text(size=30),axis.text.x=element_blank(),axis.ticks.x=element_blank())+coord_flip()+guides(fill=guide_legend(ncol =1))+scale_fill_manual(values = c('#90B30E', '#799900', '#AABA6E', '#D4D700', '#80B918', "#999900", "#555500", '#583101', '#7D3600', '#7D451B', '#826046', '#e95d00',"#F3722C", '#F8961E', '#ffbb33','#D00000','#9d0208','#6a0000','#ff0054', '#ea698b', '#C981B1', '#A35B8B','#7D3565', '#570F3F'))
  
showtext::showtext_auto()
showtext::showtext_opts(dpi = 300)
#
#ggsave("Functionalities_by_System_Focus.png", height=10, width=18, dpi=300)
```

```{r plot functionality overview}
fapro5000 %>%
  ggplot(aes(x=reorder(group, -group_total), y=group_total)) +
         geom_bar(stat="identity", fill="#347F92")+ ggtitle("Functionality Seen in >5000 Samples") + labs(y="Amount of Samples Tagged for Specified Functionality", x="Functionality Tag")+theme_light() +theme(axis.text.y=element_text(size=15), axis.title=element_text(size=35), title=element_text(size=40))+
  coord_flip()
#ggsave("Functionalities_5000.png", height=10, width=20)

fapro5000 %>%
  ggplot(aes(x=reorder(group, -group_total), y=group_total, fill=group)) +
         geom_bar(stat="identity")+ ggtitle("Functionality for by WWT Relevance") + labs(y="Amount of Samples Tagged for Specified Functionality", x="Functionality Tag")+theme_light()+theme(axis.text.y=element_text(size=15), axis.title=element_text(size=35), title=element_text(size=40), legend.text=element_text(size=20))+
  coord_flip()#+scale_fill_manual(values=c("#347F92","#52ADC2", "#F58609")) 
#ggsave("WWT_Functionalities_highlight.png", height=10, width=20)

wwt_fapro %>%
  ggplot(aes(x=reorder(group, -group_total), y=group_total)) +
         geom_bar(stat="identity", fill="#F58609")+ ggtitle("10 Most Common WWT Functional Terms") + labs(y="Amount of Samples Tagged for Specified Functionality", x="Functionality Tag")+theme_light()+theme(axis.text.y=element_text(size=30), axis.title=element_text(size=35), title=element_text(size=40))+
  coord_flip()
#ggsave("10_top_WWT_Functionalities.png", height=10, width=20)
```

```{r compairing system type, system focus, and media type}
s = sum(fullset$group_total)
ggplot(fullset, aes(x=System_Type, y=(group_total/s))) +
  geom_col(aes(fill=System_Type), position = 'dodge') + ggtitle("Function vs. System Type") + labs(x="System Type", y="Percentage")+scale_fill_manual(values=System_Colors)

s = sum(wwt$group_total)
ggplot(wwt, aes(x=System_Type, y=(group_total/s))) +
  geom_col(aes(fill=System_Type), position = 'dodge') + ggtitle("WWT Function vs. System Type") + labs(x="System Type", y="Percentage")+scale_fill_manual(values=System_Colors)

ggplot(wwt, aes(x=System_Name, y=(group_total/s))) +
  geom_col(aes(fill=System_Type), position = 'stack') + ggtitle("WWT Function vs. System Name") + labs(x="System Name", y="Percentage") +scale_fill_manual(values=System_Colors)

ggplot(wwt, aes(x=System_Name, y=(group_total/s))) +
  geom_col(aes(fill=Media_Type_1), position = 'dodge') + ggtitle("WWT Function vs. Media Type") + labs(x="Media Type", y="Percentage") +scale_fill_manual(values=Media_Colors)


## Originally these had Parcu in place of System_Focus, but it turned out the way that was generated was giving faulty data. Unsure if graphing parcu in this way would be meaningful due to small dataset.
wwt %>% arrange(Media_Type_1) %>% mutate(Media_Type_1=factor(Media_Type_1, levels=Media_Order)) %>% ggplot(aes(x=System_Focus, y=(group_percent))) +
  geom_col(aes(fill=Media_Type_1), position = 'dodge') + ggtitle("WWT Function") + labs(x="WWT System Type", y="Percentage")+theme_light() +scale_fill_manual(values=c(Media_Colors))

wwt %>% arrange(System_Type) %>% mutate(System_Type=factor(System_Type, levels=System_Order)) %>% ggplot(aes(x=System_Focus, y=group_percent, fill=System_Type)) + geom_col(aes(), position = 'dodge')+ ggtitle("WWT Targeted Functionality") + labs(y="Percent of Functional Tags")+theme_light()+scale_fill_manual(values=System_Colors)
```

```{r plot chemoheterotrophy}
# This can be changed to focus on any column in the dataset, created to zoom in on individual functional tags
fullset$Media_Type_1 = factor(fullset$Media_Type_1, levels = c("Real wastewater", "Synthetic wastewater", "Fermented dairy manure"))

fullset %>% group_by(Collection_Date) %>% ggplot(aes(x=System_Type, y=as.numeric(as.character(chemoheterotrophy)), color=Collection_Date)) +
  geom_point()

fullset %>% ggplot(aes(x=Media_Type_1, y=as.numeric(as.character(chemoheterotrophy)), fill=Media_Type_1)) + geom_col() + scale_fill_manual(values=Media_Colors)
```

```{r data processing: system type}
fullscale = fullset %>% filter(System_Type == "Full scale WRRF")
model = fullset %>% filter(System_Type == "Scale model WRRF")
batch = fullset %>% filter(System_Type == "Sequencing batch reactor")
```

```{r data processing: media type}
realwater = fullset %>% filter(Media_Type_1 == "Real wastewater")
dairywater = fullset %>% filter(Media_Type_1 == "Fermented dairy manure")
synthwater = fullset %>% filter(Media_Type_1 == "Synthetic wastewater")
```

The following sections all focus on functional tags. Ie the 'nitr' section graphs all tags with 'nitr' in them such as nitrate, nitrite, nitrogen_respiration, etc.

```{r subset aerobic}
aerobic_set = dplyr::select(fullset, contains("aerobic"))
aerobic_set$Sample_ID = fullset$Sample_ID
aerobic_set$group_total = rowSums(aerobic_set[,1:4])
aerobic_set = left_join(aerobic_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
aerobic_set_fullscale = aerobic_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
aerobic_set_model = aerobic_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
aerobic_set_batch = aerobic_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

aerobic_set_percent = rbind(aerobic_set_batch, aerobic_set_model, aerobic_set_fullscale)

ggplot(aerobic_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("Aerobic- Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)
```

```{r subset nitr}

nitr_set = dplyr::select(fullset, contains("nitr"))
nitr_set$Sample_ID = fullset$Sample_ID
nitr_set$group_total = rowSums(nitr_set[,1:13])
nitr_set = left_join(nitr_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
nitr_set_fullscale = nitr_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
nitr_set_model = nitr_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
nitr_set_batch = nitr_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

nitr_set_percent = rbind(nitr_set_batch, nitr_set_model, nitr_set_fullscale)

ggplot(nitr_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("Nitr- Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)
```

```{r subset ox}

ox_set = dplyr::select(fullset, contains("ox"))
ox_set$Sample_ID = fullset$Sample_ID
ox_set$group_total = rowSums(ox_set[,1:23])
ox_set = left_join(ox_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
ox_set_fullscale = ox_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
ox_set_model = ox_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
ox_set_batch = ox_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

ox_set_percent = rbind(ox_set_batch, ox_set_model, ox_set_fullscale)

ggplot(ox_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("Ox- Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)

```

```{r subset trophy}

trophy_set = dplyr::select(fullset, contains("trophy"))
trophy_set$Sample_ID = fullset$Sample_ID
trophy_set[,1:13] <- sapply(trophy_set[,1:13], as.numeric)
trophy_set$group_total = rowSums(trophy_set[,1:13])
trophy_set = left_join(trophy_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
trophy_set_fullscale = trophy_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
trophy_set_model = trophy_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
trophy_set_batch = trophy_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

trophy_set_percent = rbind(trophy_set_batch, trophy_set_model, trophy_set_fullscale)

ggplot(trophy_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("-trophy Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)


```

```{r subset dark}
dark_set = dplyr::select(fullset, contains("dark"))
dark_set$Sample_ID = fullset$Sample_ID
dark_set$group_total = rowSums(dark_set[,1:7])
dark_set = left_join(dark_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
dark_set_fullscale = dark_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
dark_set_model = dark_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
dark_set_batch = dark_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

dark_set_percent = rbind(dark_set_batch, dark_set_model, dark_set_fullscale)

ggplot(dark_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("Dark Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)

```

```{r subset redu}
redu_set = dplyr::select(fullset, contains("redu"))
redu_set$Sample_ID = fullset$Sample_ID
redu_set[,1:6] <- sapply(redu_set[,1:6], as.numeric)
redu_set$group_total = rowSums(redu_set[,1:6])
redu_set = left_join(redu_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
redu_set_fullscale = redu_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
redu_set_model = redu_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
redu_set_batch = redu_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

redu_set_percent = rbind(redu_set_batch, redu_set_model, redu_set_fullscale)

ggplot(redu_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("Redu- Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)

```

```{r subset resp}
resp_set = dplyr::select(fullset, contains("resp"))
resp_set$Sample_ID = fullset$Sample_ID
resp_set$group_total = rowSums(resp_set[,1:12])
resp_set = left_join(resp_set, meta, by="Sample_ID")

s = sum(fullscale$group_total)
resp_set_fullscale = resp_set %>% filter(System_Type == "Full scale WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(model$group_total)
resp_set_model = resp_set %>% filter(System_Type == "Scale model WRRF") %>% mutate(group_total_percent = group_total/s)

s = sum(batch$group_total)
resp_set_batch = resp_set %>% filter(System_Type == "Sequencing batch reactor") %>% mutate(group_total_percent = group_total/s)

resp_set_percent = rbind(resp_set_batch, resp_set_model, resp_set_fullscale)

ggplot(resp_set_percent, aes(x=System_Type, y=group_total_percent, fill=System_Type)) +
  geom_col(aes(), position = 'dodge') + ggtitle("Resp- Function vs. System Type") + labs(x="System Type", y="Percentage") + scale_fill_manual(values=System_Colors)

```
