---
title: "clean_sludge_2"
author: "Ruben Lancaster"
date: "`r Sys.Date()`"
output: html_document
---

```{r include = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
select = dplyr::select
```

### Software Used

R 4.3.1 was used for all computations. The following packages were used:

| Package | Use |
|---------|-----|
| dplyr | general data processing |
| tidyverse | general data processing | 
| DT | general data processing |
| devtools | general data processing | 
| ggplot2 | general data visualization | 
| gridExtra | general data visualization | 
| ggpubr | general data visualization |
| rstatix | general data processing and visualization | 
| vegan | microbial community analysis |
| phyloseq | microbial community analysis | 
| DESeq2 | normalization |
| MicEco | Euler graphs | 
| multcomp | multiple comparisons, analysis of variance | 
| ANCOMBC | differential abundance analysis | 
| DADA2 | resolving ASVs and assigning taxonomy | 
| fantaxtic | microbial abundance graphs |

## Background
  
Activated sludge (AS) is a mixture of microbes that play an important part in waste water treatment (WWT). Living bacteria within AS remove polluting nutrients such as phosphorus and nitrogen before effluent is released back to the water cycle. This is an exploratory study focusing on taxonomic composition of these bacterial communities and their associated functional potential, suggesting the roles they may play in AS. In this project, we worked with data provided by Dr. Erik Coats and the Bioinformatics Core from the University of Idaho. The Coats lab has sampled AS from different waste water treatment setups: from a full-scale operational waste water treatment plant in Moscow, ID, to experimental sequencing batch reactors used for increasing PHA (polyhydroxyalkanoates) production which is a bioplastics precursor, to investigating more efficient EBPR (Enhanced Biological Phosphorus Removal) and nitrogen removal systems. 
  
The data span a range of years, and samples were collected from this variety of setups. Despite this variety, we can generally separate these systems by their intended specialized function: to remove phosphorus, to remove phosphorus and nitrogen, or to enhance PHA production.

## Acknowledgements

Thank you to our collaborator Dr. Erik Coats and our BGMP mentor Dr. Maxine Wren for technical support on this project. 
Data collection and analyses performed by the IIDS Genomics and Bioinformatics Resources Core at the University of Idaho were supported in part by NIH COBRE grant P30GM103324.
  
## Data details 
  
88 individual samples of 16S rRNA gene (using Eubacteria primers for the 16S V3 region) sequenced on an individual run using multiplexing.

```{r Load packages}
library(ANCOMBC)
library(DT)
library(phyloseq)
library(vegan)
library(devtools)
library(MicEco)
library(ggplot2)
library(tidyverse)
library(gridExtra)
library(multcomp)
library(dplyr)
```

```{r set seed}
set.seed(100)
```

```{r Data ingest}

#Read in ASV file# 
asv <- read_tsv("C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/asv_table_leyla_1-9-23.tsv", show_col_types = FALSE)
asv = asv %>% column_to_rownames(var = "ASV")
# filter ASVs with only 1 occurrence
asv = as.data.frame(asv) %>% mutate(count = rowSums(.)) %>% dplyr::select(count, everything()) %>% filter(count > 1) %>% dplyr::select(-count)
#reformat to required format
asv = t(asv)
sequences = colnames(asv)

#Read in metadata file # 
meta <- read_csv("C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/Illumina_Reads_updated_1-12-24.csv", show_col_types = FALSE)
# clean metadata file
names(meta)<-make.names(names(meta),unique = TRUE) #convert column name spaces to .
meta$Collection.Date <- as.Date(meta$Collection.Date, format = "%m/%d/%Y") #format as date
#meta$Year <- as.numeric(format(meta$Collection.Date, "%Y")) #Add a year column if desired. 
meta = meta %>% dplyr::select(Sample.ID, System.Name, System.Type, Media.Type.1, Bin, Collection.Date) # remove extra metadata variables, retain only the ones we need for analysis. 

# Organize levels
mylevels = c("Enrichment BSC", "Full Enrichment BSC", "Production Run BSC", "AE4","AE8", "AE12", "AE20", "BP_PX2", "BP_PX2.1", "BP_PX3", "BP_PX3.1", "BP_PX5", "BP_PX6", "FEBPR", "GEBPR", "NF", "SEBPR", "Winkler I", "Winkler J","MWWTP", "scalemodel")
meta$System.Name = factor(meta$System.Name, levels = mylevels)
meta$Bin = sub("PHA bioplastics reactor", "PHA", meta$Bin)
meta$Bin = sub("N and P removal", "PN", meta$Bin)
meta$Bin = sub("P removal", "P", meta$Bin)
meta$Bin = factor(meta$Bin, levels = c("PN", "P", "PHA"))

# Read in tax file, if already created
tax = read_csv("C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/taxa_silva_dada2_20240110.csv", show_col_types = FALSE)
tax = tax %>% filter(...1 %in% sequences)

# Read in functional tags file, if already created
fapro_raw = read_tsv("\\\\wsl.localhost/Ubuntu/home/ruben/bioinfo/sludge/silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv", show_col_types = FALSE) %>% column_to_rownames(var = "group") %>% as.matrix()
```

```{r DADA2 assign taxonomy}

library(dada2)
# Information on using DADA2 to assign taxonomy can be found here: 
# https://benjjneb.github.io/dada2/assign.html

# vector of all the ASV sequences
seqs = colnames(asv)

#Read in path to taxonomy database of choice
silva <- "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/silva_nr99_v138.1_train_set.fa.gz"

# midas <- "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/DADA2_File_MiDAS_5.1.fa"

taxa_silva <- assignTaxonomy(seqs, silva, minBoot = 50, multithread=FALSE) #Set multithread to TRUE for faster computation if you are on a Mac.

# taxa_midas <- assignTaxonomy(seqs, midas, minBoot = 50, multithread=FALSE) #Set multithread to TRUE for faster computation if you are on a Mac.

taxa_silva_df = as.data.frame(taxa_silva)
write.csv(taxa_silva_df, "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/taxa_silva_dada2_20240110.csv", row.names=TRUE)

# taxa_midas_df = as.data.frame(taxa_midas)
# write.csv(taxa_midas_df, "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/taxa_midas_dada2.csv", row.names=TRUE)

tax = read_csv("C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/taxa_silva_dada2_20240110.csv", show_col_types = FALSE)
```

```{r create phyloseq object}
#preparation to make things ready to put into a phyloseq object 

#make names match between asv and meta files, as needed
# asv$Sample = sub("(*_F_filt.fastq.gz)", "", x = asv$Sample, ignore.case = FALSE, perl = FALSE, fixed = FALSE, useBytes = FALSE)

#columns to rownames as needed
# asv <- asv %>% column_to_rownames("Sample")
meta <- meta %>% column_to_rownames("Sample.ID")
tax <- tax %>% as.data.frame() %>% column_to_rownames(var = "...1")

#Transpose asv df to have ASVs as the rows
asv <- t(asv)

#transform to matrices 
asv <- as.matrix(asv)
tax <- as.matrix(tax)

#make phyloseq object
ASV = otu_table(asv, taxa_are_rows = TRUE)
TAX = tax_table(tax)
SAMPLE = sample_data(meta)
phyobj_all = phyloseq(ASV, TAX, SAMPLE)

#make a backup in case 
backup_phyobj_all = phyobj_all

#organize levels for system names to order them for plotting later
mylevels = c("Enrichment BSC", "Full Enrichment BSC", "Production Run BSC", "AE4","AE8", "AE12", "AE20", "BP_PX2", "BP_PX2.1", "BP_PX3", "BP_PX3.1", "BP_PX5", "BP_PX6", "FEBPR", "GEBPR", "NF", "SEBPR", "Winkler I", "Winkler J","MWWTP", "scalemodel")
sample_data(phyobj_all)$System.Name <- factor(sample_data(phyobj_all)$System.Name,
                                      levels=mylevels)
```

## Data exploration

We can evaluate the overall alpha diversity of all the samples to get a general understanding of the spread of alpha diversity. Here are diversity values using richness, Shannon's, and Simpson's index calculations from the vegan package. 

```{r Overall alpha diversity evaluation by sample}

# Alpha diverity evaluation following a Riffomonas Project tutorial from: 
# https://www.youtube.com/watch?v=wq1SXGQYgCs&ab_channel=RiffomonasProject

# prepare data into long format 
asv_df = as.data.frame(asv)
asv_names = tibble::rownames_to_column(asv_df, "ASV")
long_asv = pivot_longer(asv_names,
                        cols = 2:89,
                        names_to = "Group")
long_asv = dplyr::rename(long_asv, "name" = "ASV")
shared = long_asv

#alpha diversity functions
summary_shared = shared %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            n = sum(value))

p1 = summary_shared %>% 
  pivot_longer(cols = c(richness, shannon, simpson),
               names_to = "metric") %>% 
  ggplot(aes(x=Group, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 3, scales = "free_y")

p1
#Save image
ggsave("summary_sample_diversity.png", plot=p1, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
```

To evaluate how sampling effort affects alpha diversity, we can use random subsampling. We have a few samples (under about 10,000) that show low sampling effort may affect their results. For the most part, points follow the flat line and show that we have sufficient sampling. We should consider a cutoff of 10,000.

```{r Rarefaction: evaluate how sample size affects diversity metrics}
#code from following the tutorial available at https://riffomonas.org/code_club/2022-03-17-alpha-diversity and  https://www.youtube.com/watch?v=wq1SXGQYgCs&ab_channel=RiffomonasProject
#generate random dataframe: random subsampling 

rand <- shared %>% 
  uncount(value) %>% 
  mutate(name = sample(name)) %>% 
  dplyr::count(Group, name, name = "value")

summary_rand = rand %>% 
  group_by(Group) %>% 
  summarize(richness = specnumber(value),
            shannon = diversity(value, index = "shannon"),
            simpson = diversity(value, index = "simpson"),
            n = sum(value))

p1= summary_rand %>% 
  pivot_longer(cols = c(richness, shannon, simpson),
               names_to = "metric") %>% 
  ggplot(aes(x=n, y=value)) + #plot by sample size
  geom_point()+
  geom_smooth()+
  facet_wrap(~metric, nrow = 4, scales = "free_y")+
  ggtitle("Random sampling")+
  xlab("Sample size")

p1
#Save image
ggsave("summary_rarified_diversity.png", plot=p1, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
```

```{r data preprocessing}

phyobj_all = backup_phyobj_all
phyobj_all

#Subset out other Kingdoms (one Eukaryote), mitochondria, and chloroplasts which are picked up by the eubacteria primer


#prune to 10,000. See alpha diversity section.
samples = sample_sums(phyobj_all) > 10000
phyobj_all = prune_samples(samples, phyobj_all) 

phyobj_all = subset_taxa(phyobj_all, Kingdom == "Bacteria")
phyobj_all = subset_taxa(phyobj_all, Family != "Mitochondria")
phyobj_all = subset_taxa(phyobj_all, Order != "Chloroplast")


# List the samples that have been removed in preprocessing
new_table = otu_table(phyobj_all) %>% as.data.frame()
new_samples = colnames(new_table)
old_samples = rownames(meta)

diff = unique(old_samples[! old_samples %in% new_samples]) 
diff

phyobj_all

```

## Data preprocessing 

We performed the following data preprocessing steps:
1a. Remove any reads not in the kingdom Bacteria
1b. Remove reads tagged as mitochondria
1c. Remove reads tagged as chloroplasts
2. Remove samples with less than 10,000 reads
3. Remove ASVs that only appear once. These are possible sequencing errors. 


```{r taxonomic resolution}
# Check for best resolution available from taxonomic assignment. #

#Before filtering
tax_df = tax_table(backup_phyobj_all) %>% as.data.frame()

sum(is.na(tax_df$Class))/length(tax_df$Class)
sum(is.na(tax_df$Order))/length(tax_df$Order)
sum(is.na(tax_df$Family))/length(tax_df$Family)
sum(is.na(tax_df$Genus))/length(tax_df$Genus)

# After filtering
tax_df = tax_table(phyobj_all) %>% as.data.frame()

sum(is.na(tax_df$Class))/length(tax_df$Class)
sum(is.na(tax_df$Order))/length(tax_df$Order)
sum(is.na(tax_df$Family))/length(tax_df$Family)
sum(is.na(tax_df$Genus))/length(tax_df$Genus)
```
Before filtering, Order provides the highest taxonomic resolution available (~7% NA), with the next taxonomic level of Family not providing enough specificity (~22% NA). After filtering, Family provides the highest taxonomic resolution available (31% of taxa not identified to the genus level).

## Alpha Diversity Analysis

Here we use Shannon’s index, which considers both richness and evenness and tends to outperform Simpson’s index when many rarely occurring species are present, which is suitable for our dataset. 

```{r Shannon's by Bin type, fig.show = "hide"}
#ANOVA
richness = phyobj_all %>% estimate_richness(measures = c("Shannon", "Simpson")) %>% rownames_to_column(var="Sample.ID")
  #subset meta 
samples = unique(richness$Sample.ID)
meta_subset = meta %>% rownames_to_column(var="Sample.ID") %>% filter(Sample.ID %in% samples)

diversity = full_join(meta_subset,richness, by="Sample.ID")

diversity$Bin = diversity$Bin %>% as.character() %>% as.factor()
shannon_bin = aov(Shannon ~ Bin - 1, data = diversity)

#Check Assumptions: ANOVA
qqnorm(shannon_bin$residuals)
qqline(shannon_bin$residuals)

data.frame(
  Media = diversity$Bin,
  Residuals = shannon_bin$residuals
) %>% 
  ggplot(aes(x = Media, y = Residuals, fill = Media)) +
  geom_boxplot()+
  ggtitle("Shannon's: Bin Type")

#Pairwise Comparisons with Tukey's Method
summary(glht(shannon_bin, linfct = mcp(Bin = "Tukey")))

```

```{r Alpha Diversity Analysis: figures}
library(ggpubr)
library(rstatix)

#Use rstatix package for plotting ggplot: slightly different results for Tukey Post-Hoc but same significance
#https://github.com/kassambara/ggpubr/issues/102#issuecomment-411897608

stat.test <- shannon_bin %>% tukey_hsd()
stat.test$custom.label <- ifelse(stat.test$p.adj <= 0.05, stat.test$p.adj, "ns")

p1 = plot_richness(phyobj_all, measures = c("Shannon"), x= "Bin")+
  xlab("") + theme(axis.text.x = element_text(angle = 0, hjust = 0.5, size = 12))+
    stat_pvalue_manual(stat.test, label = "p.adj.signif", y.position = c(6, 7, 8), tip.length = 0.01)

p1
#Save plot
ggsave(filename = "shannons_bin.png", plot = p1, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
```

Nitrogen and phosphorus removing systems have significantly different alpha diversity measures from phosphorus removing systems and PHA bioreactors (ANOVA with Tukey HSD, p < 0.00001). PHA bioplastics reactors and phosphorus removing systems do not have significantly different alpha diversity measures (ANOVA with Tukey HSD, p = 0.952). 
  
## Community structure with Euler diagrams 

```{r MicEco Euler diagrams, Full Sample}
library(MicEco)
library(gridExtra)

set.seed(130) #it looks really wonky otherwise.
euler1 = ps_euler(phyobj_all, "Media.Type.1", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#F2D000", "#F58609", "#A01892"))
p1 = grid.arrange(grobs = list(euler1), top = "Media Type, Full Sample")

set.seed(100)
euler2 = ps_euler(phyobj_all, "System.Type", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#347F92", "#A6CEE3","#94DE11" ))
p2 = grid.arrange(grobs = list(euler2), top = "System Type, Full Sample")

set.seed(123)
euler3 = ps_euler(phyobj_all, "Bin", quantities = list(type=c("percent")), fraction = 0, shape = "ellipse", weight = TRUE, relative = TRUE, fill = c("#347F92", "#A6CEE3","#94DE11" ))
p3 = grid.arrange(grobs = list(euler3), top = "Bin, Full Sample")

ggsave(filename = "euler_media.png", plot = p1, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
ggsave(filename = "euler_system.png", plot = p2, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
ggsave(filename = "euler_bin.png", plot = p3, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
```


## Beta Diversity analysis
  
We performed variance stabilization to correct for different sequencing depth or amplification bias before performing beta diversity analysis using the DESeq2 package. 

```{r preprocessing, variance stabilization (normalization)}

# Initial code courtesy of Dr. Maxine Wren, modified to fit our analysis

library(DESeq2)

#First, we need to create a DESeq2 object. Selected experimental variable is Bin (expected function)
diagdds <- phyloseq_to_deseq2(phyobj_all, ~Bin)

#Next, we need to estimate the variance between samples
diagdds <- estimateSizeFactors(diagdds,type="poscounts")
diagdds <- estimateDispersions(diagdds) 

#Now we need to use the variance between samples to normalize the data. 
#This reduces "noise" from variation between samples that may be due to differences
#in amplification or sequencing depth.
diagvst <- getVarianceStabilizedData(diagdds) 

#Now we need to make it BACK into an ASV/OTU table and slot it back into the phyloseq.
otu_table(phyobj_all) <- otu_table(diagvst, taxa_are_rows = TRUE)
ot_df <- data.frame(otu_table(phyobj_all))
ot_df[ot_df < 0.0] <- 0.0 # why this?
ot <- otu_table(ot_df, taxa_are_rows = TRUE)
sample_names(ot) <- sample_names(phyobj_all@otu_table)
otu_table(phyobj_all) <- otu_table(ot, taxa_are_rows = TRUE)

backup_var_phyobj_all = phyobj_all
```

```{r Functionality: Faprotax}
#Prepare a properly formatted tax table for use with Faprotax.

fasv = otu_table(phyobj_all) %>% as.data.frame() %>% rownames_to_column("otu")
ftax = tax_table(phyobj_all) %>% as.data.frame()

ftax$Kingdom = paste("d", ftax$Kingdom, sep = "__")
ftax$Phylum = paste("p", ftax$Phylum, sep = "__")
ftax$Class = paste("c", ftax$Class, sep = "__")
ftax$Order = paste("o", ftax$Order, sep = "__")
ftax$Family = paste("f", ftax$Family, sep = "__")
ftax$Genus = paste("g", ftax$Genus, sep = "__")
ftax$Species = paste("s", ftax$Species, sep = "__")

ftax$Taxonomy = paste(ftax$Kingdom, ftax$Phylum, ftax$Class, ftax$Order, ftax$Family, ftax$Genus, ftax$Species, sep = ";")

tax_table = ftax %>% rownames_to_column(var = "otu") %>% dplyr::select(otu, Taxonomy)


export_tax_table = dplyr::left_join(x = as.data.frame(tax_table),
                             y = as.data.frame(fasv),
                             by = "otu") %>% dplyr::select(-otu)
 
write_tsv(export_tax_table, "silva_allgroups_taxtable_1-27-24-filt-norm.tsv")

#Run Faprotax on the command line. You will need numpy installed. 
# ./collapse_table.py -i silva_allgroups_taxtable_1-27-24-filt-norm.tsv -g FAPROTAX.txt -o silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv -r silva_allgroups_taxtable_1-27-24-filt-norm_report.txt -d "Taxonomy" -v

fapro_raw = read_tsv("\\\\wsl.localhost/Ubuntu/home/ruben/bioinfo/sludge/silva_allgroups_taxtable_1-27-24-filt-norm_out.tsv", show_col_types = FALSE) %>% column_to_rownames(var = "group") %>% as.matrix()
```

## Relative Abundance

We can visualize the diversity of each individual sample, each system, and each operation type at the class level. 

```{r Relative Abundance: by Class}

# Following the phyloseq relative abundance tutorial posted here: 
# https://github.com/joey711/phyloseq/issues/1701 

#subset only Bacteria
normalized_ps = phyobj_all

#select your variable of interest here 
merged_ps_phylum_SysName = merge_samples(normalized_ps, "System.Name") 
merged_ps_phylum_Bin = merge_samples(normalized_ps, "Bin")
# merged_ps_phylum_Media = merge_samples(normalized_ps, "Media.Type.1") 
# merged_ps_phylum_SysType = merge_samples(normalized_ps, "System.Type") 

#Merge at your taxonomic rank of interest, I chose Class here.
normalized_ps_phylum1 = tax_glom(normalized_ps, "Class", NArm = TRUE)
normalized_ps_phylum2 = tax_glom(merged_ps_phylum_SysName, "Class", NArm = TRUE)
normalized_ps_phylum3 = tax_glom(merged_ps_phylum_Bin, "Class", NArm = TRUE)
# normalized_ps_phylum4 = tax_glom(merged_ps_phylum_Media, "Class", NArm = TRUE)
# normalized_ps_phylum5 = tax_glom(merged_ps_phylum_SysType, "Class", NArm = TRUE)

# Transform counts to relative abundances
normalized_ps_phylum1_relabun = transform_sample_counts(normalized_ps_phylum1, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum2_relabun = transform_sample_counts(normalized_ps_phylum2, function(OTU) OTU/sum(OTU) * 100)
normalized_ps_phylum3_relabun = transform_sample_counts(normalized_ps_phylum3, function(OTU) OTU/sum(OTU) * 100)
# normalized_ps_phylum4_relabun = transform_sample_counts(normalized_ps_phylum4, function(OTU) OTU/sum(OTU) * 100)
# normalized_ps_phylum5_relabun = transform_sample_counts(normalized_ps_phylum5, function(OTU) OTU/sum(OTU) * 100)

#Get the number of unique colors and select that many from iWantHue https://medialab.github.io/iwanthue/ 
tax_df = tax %>% as.data.frame()
length(unique(tax_df$Class))

myColors <- c("#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549")

# assign specific colors to variables
# https://stackoverflow.com/questions/6919025/how-to-assign-colors-to-categorical-variables-in-ggplot2-that-have-stable-mappin
tax_df = tax %>% as.data.frame()
names(myColors) <- levels(tax_df$Class)


p1 = psmelt(normalized_ps_phylum1_relabun) 
p1 = p1 %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Sample", x = "") +
        guides(fill = "none") + scale_fill_manual(name = "Class", values=myColors)

p2 = psmelt(normalized_ps_phylum2_relabun) 
p2$Sample = factor(p2$Sample, levels= mylevels)
p2 = p2 %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by System", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors)

p3 = psmelt(normalized_ps_phylum3_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Class Relative Abundance by Bin", x = "") +
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) 

# p4 = psmelt(normalized_ps_phylum4_relabun) %>%
#         ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
#         geom_bar(stat = "identity") +
#         theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#         labs(y = "Relative Abundance", title = "Class Relative Abundance", x = "") +
#         guides(fill="none")+ scale_fill_manual(name = "Class", values=myColors)
# 
# p5 = psmelt(normalized_ps_phylum5_relabun) %>%
#         ggplot(aes(x = Sample, y = Abundance, fill = Class)) +
#         geom_bar(stat = "identity") +
#         theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#         labs(y = "Relative Abundance", title = "Class Relative Abundance", x = "") +
#         guides(fill="none") + scale_fill_manual(name = "Class", values=myColors)


p1 
p2 
p3 + theme_bw()
# p4
# p5


#Save plots
ggsave(filename = "relabund_class_sampleid.png", plot = p1, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
ggsave(filename = "relabund_class_samplename.png", plot = p2, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
```

We can also visualize the diversity of functions present in each system. 

```{r Relative Abundance: by functionality}
#Filter out subgroupings of functionality
# fapro_filt = fapro_raw %>% t() %>% as.data.frame() %>%  dplyr::select(-c(oxygenic_photoautotrophy, methanol_oxidation, dark_sulfide_oxidation, aerobic_nitrite_oxidation, human_gut, human_pathogens_all, mammal_gut, anoxygenic_photoautotrophy, anoxygenic_photoautotrophy_S_oxidizing, photosynthetic_cyanobacteria, photoheterotrophy, thiosulfate_respiration, methanotrophy, aliphatic_non_methane_hydrocarbon_degradation, iron_respiration, aerobic_chemoheterotrophy, nitrite_respiration, photoautotrophy, sulfur_respiration, nitrate_denitrification, nitrite_denitrification, nitrous_oxide_denitrification, denitrification, nitrate_respiration, nitrogen_respiration, human_associated, aromatic_hydrocarbon_degradation, chloroplasts)) %>% t()

func_ASV = otu_table(fapro_raw, taxa_are_rows = TRUE)
phyobj_func = phyloseq(func_ASV, SAMPLE)

# plot all functions by each individual sample
merged_ps_phylum_SysName = merge_samples(phyobj_func, "System.Name") 
ps_phylum1_relabun = transform_sample_counts(merged_ps_phylum_SysName, function(OTU) OTU/sum(OTU) * 100)
myColors <- c("#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549","#645fe8", "#6ebf25", "#872bb2", "#38d65f", "#a725ae", "#51c33b", "#cd6af2", "#009a1a", "#d10f9a", "#4ee084", "#f21174", "#01d6a8", "#da1932", "#01d0ba", "#c9003a", "#009946", "#ff78ee", "#247f00", "#0262e8", "#abb900", "#533fb3", "#9cd763", "#807dff", "#ffa331", "#015bb2", "#e8690c", "#1ebdff", "#cf4301", "#b091ff", "#898600", "#5f419a", "#5e7600", "#ff99f0", "#186012", "#c7006c", "#90d792", "#8f2871", "#cacb75", "#ff61a3", "#007c54", "#e24224", "#008d8d", "#ff914a", "#94bcff", "#bc7100", "#e0b1ff", "#9b7900", "#ffabef", "#325c2c", "#ffa8c9", "#5c5312", "#c492be", "#b75c00", "#7cc9af", "#962b3b", "#e5c276", "#72426a", "#f6ba8e", "#922e4f", "#6b4e00", "#ff8e98", "#854900", "#ff9182", "#7f4a34", "#ff9972", "#8e371c", "#d38f81", "#996549")
p = psmelt(ps_phylum1_relabun) %>%
        ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Relative abundances of tagged functions", x = "") + 
        guides(fill="none") + scale_fill_manual(name = "Class", values=myColors) 
p
ggsave(filename = "allfunctions_systemname.png", plot = p, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")

# get top functions for bins

func_ASV = otu_table(fapro_raw, taxa_are_rows = TRUE)
phyobj_func = phyloseq(func_ASV, SAMPLE)

ps = phyobj_func

p = subset_samples(ps, Bin == "P")
p10 = names(sort(taxa_sums(p), TRUE)[1:10])
pn = subset_samples(ps, Bin == "PN")
pn10 = names(sort(taxa_sums(pn), TRUE)[1:10])
pha = subset_samples(ps, Bin == "PHA")
pha10 = r10 = names(sort(taxa_sums(pha), TRUE)[1:10])
top10 = unique(c(pha10, p10, pn10))

ps = prune_species(top10, ps)

merged_ps = merge_samples(ps, "Bin") #change what variable here

merged_ps = transform_sample_counts(merged_ps, function(OTU) OTU/sum(OTU) * 100)

#get the number of unique functions and select that many colors as above

myColors <- c("#006403", "#9d55dd", "#d5b500", "#88258e", "#a9d382", "#b0008a", "#d04b00", "#0082d2", "#ff635a", "#e99eff", "#814030", "#fcb0d3", "#9c1b5c", "firebrick", "#a0029b", "seagreen")
# tax_df = otu_table(ps) %>% as.data.frame()
# names(myColors) <- rownames(tax_df)

p1 = psmelt(merged_ps) %>% 
  ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
        geom_bar(stat = "identity") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
        labs(y = "Relative Abundance", title = "Most common functions", x = "") +
        scale_fill_manual(values=myColors) + 
        guides(fill=guide_legend(title="Function")) + theme_bw()
p1
ggsave(filename = "topfunction_bin.png", plot = p1, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")

#get top functions for media type
# ps = phyobj_func
# synthetic = subset_samples(ps, Media.Type.1 == "Synthetic wastewater")
# s10 = names(sort(taxa_sums(synthetic), TRUE)[1:10])
# raw = subset_samples(ps, Media.Type.1 == "Real wastewater")
# r10 = names(sort(taxa_sums(raw), TRUE)[1:10])
# dairy = subset_samples(ps, Media.Type.1 == "Fermented dairy manure")
# d10 = r10 = names(sort(taxa_sums(dairy), TRUE)[1:10])
# top10 = unique(c(s10, r10, d10))
# 
# ps = prune_species(top10, ps)
# 
# merged_ps = merge_samples(ps, "Media.Type.1") #change what variable here
# 
# merged_ps = transform_sample_counts(merged_ps, function(OTU) OTU/sum(OTU) * 100)
# #get the number of unique functions and select that many colors as above
# 
# p2 = psmelt(merged_ps) %>% 
#   ggplot(aes(x = Sample, y = Abundance, fill = OTU)) +
#         geom_bar(stat = "identity") +
#         theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
#         labs(y = "Relative Abundance", title = "Most common functions", x = "") +
#         scale_fill_manual(values=myColors) + 
#         guides(fill=guide_legend(title="Function")) + theme_bw()
# 
# p2


#Save plot
# ggsave(filename = "topfunction_mediatype.png", plot = p2, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")
```

We can see exactly what taxa are the most represented in these systems. 

```{r fantaxtic top genera}
# Top taxa package for plotting nested ggplots from a phyloseq object, fantaxtic, is here: 
# https://github.com/gmteunisse/Fantaxtic

#clear any merging and changing done to phyobj_all in last chunks
# phyobj_all = backup_var_phyobj_all

library(fantaxtic)
library(magrittr)
library(ggnested)
library(gridExtra)

#Find top 10 Orders across all samples
phyobj_all = backup_var_phyobj_all
top_asv <- top_taxa(phyobj_all, n_taxa = 10, tax_level = "Order")
p = plot_nested_bar(ps_obj = top_asv$ps_obj,
                top_level = "Class",
                nested_level = "Order") +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black"))+
  xlab("") +
  ggtitle("Top 10 Orders across all samples")
p
ggsave(filename = "topfunction_order_all.png", plot = p, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")

# Now, top 10 by separated by Bin
phyobj_all = backup_var_phyobj_all
norm_ps = merge_samples(phyobj_all, "Bin")
top_asv <- top_taxa(norm_ps, n_taxa = 10, tax_level = "Order")
p = plot_nested_bar(ps_obj = top_asv$ps_obj,
                top_level = "Class",
                nested_level = "Order")+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black"))+
  xlab("")+
  ggtitle("Top 10 Orders by Bin")
p
ggsave(filename = "topfunction_order_bin.png", plot = p, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")

# Find top Classes
top_nested <- nested_top_taxa(phyobj_all,
                              top_tax_level = "Class",
                              nested_tax_level = "Order",
                              n_top_taxa = 3, 
                              n_nested_taxa = 3)
p = plot_nested_bar(ps_obj = top_nested$ps_obj,
                top_level = "Class",
                nested_level = "Order")+
  ggtitle("Top 3 orders of top 3 classes")
p
ggsave("top_3_class_order.png", plot = p, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")

# Find top n by Variable Type 
top_grouped <- top_taxa(phyobj_all,
                        tax_level = "Order",
                        n_taxa = 5,
                        grouping = "Bin")
top_grouped$top_taxa %>% ggplot(aes(x=Bin, y = abundance*100, fill = Order)) +
  geom_bar(stat = "identity")+
  ylab("Percent abundance")

# change levels 

vars = get_variable(phyobj_all, "System.Name")
vars = factor(vars)
levels = levels(vars)
corlevels = c(levels[1:11], "Winkler I", "Winkler J", "Production Run BSC", "Full Enrichment BSC", "SEBPR", "FEBPR", "GEBPR","NF", "MWWTP", "scalemodel")
sample_data(phyobj_all)$System.Name = levels(corlevels)


# Facet wrap 
top_nested <- nested_top_taxa(phyobj_all,
                          top_tax_level = "Class",
                          nested_tax_level = "Order",
                          n_top_taxa = 10, 
                          n_nested_taxa = 1)
p = plot_nested_bar(top_nested$ps_obj,
                top_level = "Class",
                nested_level = "Order",
                legend_title = "Class and Order") +
  facet_wrap(~Bin,
             scales = "free_x",
             ncol = 6) +
  labs(title = "Relative abundances of the top 3 orders for each of the top 3 classes") +
  theme(plot.title = element_text(hjust = 0.5, 
                                  size = 20, 
                                  face = "bold"),
        legend.key.size = unit(10, 
                               "points"))
p
ggsave("top_3_classes_top5_order_bin.png", plot = p, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images", width = 8000, height = 4000, units = "px", dpi = 380)
```

```{r Top taxa poster figure generation}

# displays poorly in R markdown but saved image has correct dimensions #

#https://github.com/gmteunisse/Fantaxtic

library(extrafont)
library(showtext)
font_add(family = "skeena", regular = "\\\\wsl.localhost/Ubuntu/home/ruben/bioinfo/sludge/skeena_d.ttf")
showtext_auto()

#Find top 10 across all samples
norm_ps = merge_samples(phyobj_all, "Bin")
norm_ps@sam_data$Bin = factor(norm_ps@sam_data$Bin, levels = c("PN", "P", "PHA"))
top_asv <- top_taxa(norm_ps, n_taxa = 10, tax_level = "Order")

# options(repr.plot.width = 4, repr.plot.height =1.5) 
p3 = plot_nested_bar(ps_obj = top_asv$ps_obj,
                     top_level = "Class",
                     nested_level = "Order",
                     palette = c(Alphaproteobacteria = "#F3722C", 
                                 Bacteroidia = "#871762",
                                 Gammaproteobacteria = "#FFBB33"),
                     merged_clr = "#919191",
                     sample_order = c("PN", "P", "PHA"),
                     legend_title = "Orders by Class") +
  labs(title = "10 most represented orders") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5, color = "black", family = "skeena", size = 40),
        axis.text.y = element_text(family = "skeena", size = 40),
        title = element_text(family = "skeena", size = 50),
        plot.background = element_rect(fill = "#F0EEE4"),
        plot.title = element_text(hjust = 0.5), 
        legend.background = element_blank(),
        legend.key = element_blank(),
        legend.title = element_text(family = "skeena", size = 40),
        legend.text = element_text(size = 40, family = "skeena")) +
  xlab("")
  


p3
ggsave("top_10_orders_bin.png", plot = p3, path = "C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images", width = 7, height = 4.5, units = "in", dpi = 300)
```

```{r ANCOMBC analysis}
# Code courtesy of Dr. Maxine Wren, modified for this data

# remove collinear variables from the metadata
filtered_meta = meta %>% dplyr::select(-c("System.Type", "Media.Type.1"))
ASV = otu_table(backup_phyobj_all)
TAX = tax_table(backup_phyobj_all)
SAMPLE = sample_data(filtered_meta)
filtered_ps = phyloseq(ASV, TAX, SAMPLE)

## remove any taxa with <10 reads in the total data set (sum of all samples <10). This is because extremely rare taxa can produce a series of false positives. 

filtered_ps = prune_taxa(taxa_sums(filtered_ps) > 10, filtered_ps) 

#check the results of filtering
otu_df = otu_table(filtered_ps) %>% as.data.frame() %>% mutate(sums = rowSums(.)) %>% dplyr::select(sums, everything())

# Continuous covariate: "Collection.Date"
# Categorical covariates: “System.Name, Bin”
# The group variable of interest: “Bin”
# Three groups: "P removal", "N and P removal", 
# The reference group: “Full scale WRRF”

# refactor the categorical variables so PHA bioreactor is the comparator
sample_data(filtered_ps)$Bin <- factor(sample_data(filtered_ps)$Bin,
                                      levels=c("PHA bioplastics reactor", 
                                               "N and P removal", 
                                               "P removal"))
# If I do the fix_formula as System.Name + Collection.Date + Bin it gives me an error about checking for collinearity or missing values in my covariates, so I am just running it with "bin" as the formula.

ancombc_out = ancombc2(filtered_ps,
                             fix_formula = "Bin",
                             tax_level = "Order",
                             group = "Bin",
                             struc_zero = TRUE,
                             neg_lb = TRUE,
                             verbose = TRUE,
                             alpha = 0.05,
                             global = TRUE,
                             pairwise = TRUE,
                             dunnet = TRUE
                             )
```

```{r}
# taxon presence/absence
tab_zero = ancombc_out$zero_ind %>% column_to_rownames(var = "taxon") #%>% mutate(sum = rowSums(.)) %>% filter(sum == 2)

res_prim = ancombc_out$res
res_global= ancombc_out$res_global %>% mutate_if(is.numeric, function(x) round(x, 5))
res_global_passed = res_global %>% filter(passed_ss == TRUE) %>% filter(diff_abn == TRUE)
diff_taxons = res_global_passed$taxon
diff_taxons
```

 [1] "Bacteria_Bacteroidota_Bacteroidia_Chitinophagales"          
 [2] "Bacteria_Bacteroidota_Bacteroidia_Cytophagales"             
 [3] "Bacteria_Bacteroidota_Kapabacteria_Kapabacteriales"         
 [4] "Bacteria_Bacteroidota_Bacteroidia_Sphingobacteriales"       
 [5] "Bacteria_Proteobacteria_Alphaproteobacteria_Rhodobacterales"
 [6] "Bacteria_Planctomycetota_Planctomycetes_Pirellulales"       
 [7] "Bacteria_Chloroflexi_Anaerolineae_SBR1031"                  
 [8] "Bacteria_Proteobacteria_Alphaproteobacteria_Caulobacterales"
 [9] "Bacteria_Planctomycetota_Phycisphaerae_Phycisphaerales"     
[10] "Bacteria_Myxococcota_Polyangia_Nannocystales"               
[11] "Bacteria_Planctomycetota_Planctomycetes_Isosphaerales"      
[12] "Bacteria_Patescibacteria_Parcubacteria_NA"                  
[13] "Bacteria_Chloroflexi_Chloroflexia_Thermomicrobiales"        
[14] "Bacteria_Firmicutes_Clostridia_Oscillospirales"             
[15] "Bacteria_Proteobacteria_Alphaproteobacteria_NA"             
[16] "Bacteria_Firmicutes_Bacilli_Erysipelotrichales" 

```{r Select Parcubacteria and outgroups for use in phylogenetic tree: input format for multiple sequence alignment using MEGA}
# parcu = tax %>% as.data.frame() %>% rownames_to_column("otu") %>% filter(Class %in% c("Parcubacteria"))
# otu_tax = tax %>% as.data.frame() %>% rownames_to_column("otu")
# Gammaproteobacteria = otu_tax[8,]
# Clostridia = otu_tax[539,]
# Gracilibacteria = otu_tax[85,]
# outgroups = rbind(parcu, Gammaproteobacteria, Clostridia, Gracilibacteria)
# 
# outgroups$Kingdom = paste("d", outgroups$Kingdom, sep = "__")
# outgroups$Phylum = paste("p", outgroups$Phylum, sep = "__")
# outgroups$Class = paste("c", outgroups$Class, sep = "__")
# outgroups$Order = paste("o", outgroups$Order, sep = "__")
# outgroups$Family = paste("f", outgroups$Family, sep = "__")
# outgroups$Genus = paste("g", outgroups$Genus, sep = "__")
# outgroups$Species = paste("s", outgroups$Species, sep = "__")
# 
# outgroups$Taxonomy = paste(outgroups$Kingdom, outgroups$Phylum, outgroups$Class, outgroups$Order, outgroups$Family, outgroups$Genus, outgroups$Species, sep = ";")
# 
# outgroups = outgroups %>% dplyr::select(otu, Taxonomy)
# 
# write_tsv(outgroups, "outgroups_clostridia_gracilibacteria_gamma")

```

# Bray Curtis NMDS 

```{r variance stabilized NMDS: Bin whole sample}
set.seed(100)
#Bray-Curtis distance matrix calculation by System Type

physeq_ord <- ordinate(
 physeq = phyobj_all,
 method = "NMDS",
 formula = ~Bin,
 distance = "bray"
)
stress = as.character(round(physeq_ord$stress, digits = 3))
stresslab = paste("Stress =", stress)

p1 = plot_ordination(
 physeq = phyobj_all,
 ordination = physeq_ord,
 color = "Bin"
) +
 aes(group = Bin) +
 geom_point(aes(color = Bin), alpha = 0.95, size = 6) +
  scale_color_manual(values=c("#A6CEE3","#347F92","#94DE11"))+
 stat_ellipse(type = "t", geom = "polygon", position = "identity", alpha = .05, aes(fill = Bin)) + 
 scale_fill_manual(values=c("#A6CEE3","#347F92","#94DE11")) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
    panel.background = element_blank(), axis.line = element_line(colour = "black"),strip.text = element_text(colour = 'white'),text = element_text(size=12),legend.text=element_text(size=12),legend.title = element_blank())+
  annotate("text", x = 2, y = -2, label = stresslab) + 
  annotate("text", x = 2, y = -2.5, label = "Non-metric fit, R² = 0.985")

p1
ggsave("nmds_Bin.png", plot = p1, path="C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images")

#filter out ASVs that have only 0 values
ot_df = otu_table(phyobj_all) %>% as.data.frame()

#247 -> 239
ot_df = ot_df %>% mutate(sum = rowSums(.)) %>% dplyr::select(sum, everything()) %>% filter(sum != 0) %>% dplyr::select(-sum) %>% t()
samples = rownames(ot_df)
adonis_dist = vegdist(ot_df, method = "bray") 

# adonis_mds = metaMDS(adonis_dist) %>% scores() %>% 
#   as_tibble(rownames="Group")

#subset samples needed from meta
# needed_samples = rownames(ot_df)
# parcu_adonis_meta = meta %>% rownames_to_column(var="Group") %>% filter(Group %in% needed_samples)
# 
# meta_nmds = inner_join(adonis_mds, parcu_adonis_meta, by = "Group")

# centroid_systemtype = meta_nmds %>% 
#   group_by(Bin) %>% 
#   summarize(NMDS1 = mean(NMDS1), NMDS2 = mean(NMDS2))
# 
# meta_nmds %>% 
#   ggplot(aes(x=NMDS1, y = NMDS2, color = Bin))+
#   geom_point() +
#   stat_ellipse(show.legend = FALSE) +
#   geom_point(data = centroid_systemtype, size = 2, shape = 21, color = "black", aes(fill=Bin), show.legend = FALSE)+
#  # ggtitle("Media Type, bray-curtis")+
#   scale_colour_discrete(name="Bin") +
#   theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
#     panel.background = element_blank(), axis.line = element_line(colour = "black"))

stressplot(physeq_ord)

submeta = meta %>% t() %>% as.data.frame %>% dplyr::select(any_of(samples)) %>% t() %>% as.data.frame()

# PERMANOVA # 
bintype = adonis2(adonis_dist ~ Bin, data = submeta, permutations = 1e4)
bintype$`Pr(>F)`[1] #p-value 

bd = betadisper(adonis_dist, submeta$Bin)
anova(bd)


```

```{r NMDS by bin, for poster}
# displays poorly in R but saves image with correct dimensions # 

physeq_ord <- ordinate(
 physeq = phyobj_all,
 method = "NMDS",
 formula = ~Bin,
 distance = "bray"
)

stress = as.character(round(physeq_ord$stress, digits = 3))
stresslab = paste("Stress =", stress)

p1 = plot_ordination(
 physeq = phyobj_all,
 ordination = physeq_ord,
 color = "Bin"
) +
 aes(group = Bin) +
 geom_point(aes(color = Bin), alpha = 0.95, size = 5) +
  scale_color_manual(values=c("#F3722C","#871762","#90B30E"))+
  scale_fill_manual(values=c("#F3722C","#871762","#90B30E")) +
  stat_ellipse(type = "t", geom = "polygon", position = "identity", alpha = .05, aes(fill = Bin)) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
    panel.background = element_blank(), axis.line = element_line(colour = "black"),strip.text = element_text(colour = 'white'),text = element_text(size=60, family = "skeena"),legend.text=element_text(size=60),legend.title = element_blank())+
  annotate("text", x = 4.2, y = -2, label = stresslab, size = 12, family = "skeena")+
  annotate("text", x = 3.6, y = -2.5, label = "Non-metric fit, R² = 0.985", size = 12, family = "skeena")+
  theme(plot.background = element_rect(fill = "#F0EEE4"), legend.background = element_rect(fill = "#F0EEE4"))
  

p1
ggsave("nmds_Bin.png", plot = p1, path="C:/Users/rlanc/OneDrive/Documents/Bioinformatics/Parcubacteria_project/images", height = 5, width = 7, dpi = 300)
```

Efforts were made at the beginning of the project to investigate Parcubacteria specifically. Beta diversity analysis using NMDS proved difficult due to the sparsity of Parcubacteria differences between samples. Presence-absence and abundance testing was possible, however, and revealed that PN systems and systems with real wastewater had a greater amount and variety of Parcubacteria.

```{r Parcubacteria abundance}
# https://joey711.github.io/phyloseq/plot_bar-examples.html #

# normalize using DESeq2 first # 
phyobj_parcu = subset_taxa(phyobj_all, Class == "Parcubacteria")

p = plot_bar(phyobj_parcu, "Order", fill="Order", facet_grid=~Bin)
p + geom_bar(aes(color=Order, fill=Order), stat="identity", position="stack") + ggtitle("Parcubacteria Abundances") + ylab ("Abundance (arbitraty units)")


p = plot_bar(phyobj_parcu, "Order", fill="Order", facet_grid=~Media.Type.1)
p + geom_bar(aes(color=Order, fill=Order), stat="identity", position="stack") + ggtitle("Parcubacteria Abundances") + ylab ("Abundance (arbitraty units)")

```



